{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "02b2641f",
   "metadata": {},
   "source": [
    "#### Here we investigate the relashionship between:\n",
    "    - mRNA level predictability of a landmark gene \n",
    "    and \n",
    "    - its known organelle level biological function using GO annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "367cc88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "def locations_of_substring(string, substring):\n",
    "    \"\"\"Return a list of locations of a substring.\"\"\"\n",
    "    substring_length = len(substring)    \n",
    "    def recurse(locations_found, start):\n",
    "        location = string.find(substring, start)\n",
    "        if location != -1:\n",
    "            return recurse(locations_found + [location], location+substring_length)\n",
    "        else:\n",
    "            return locations_found\n",
    "    return recurse([], 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a60cc596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', 'DNA-directed RNA polymerase,DNA-directed RNA polymerase,'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import sys\n",
    "# sys.path.insert(0, '../utils/') \n",
    "# from readProfiles import rename_affyprobe_to_genename\n",
    "# dfcats2,_=rename_affyprobe_to_genename(dfcats.set_index('ID').T,dfcats.ID.tolist())\n",
    "# dfcats2.T.reset_index()\n",
    "\n",
    "gene_cats_bpcc['UP_KW_CELLULAR_COMPONENT'].astype(str).apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('KW')[0] for si in locations_of_substring(x,c)]) if c in x else '').unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c58760cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ER'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c='DNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b1cf8192",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                                 KW-0496~Mitochondrion,\n",
       "1                                                    NaN\n",
       "2      KW-0256~Endoplasmic reticulum,KW-0472~Membrane...\n",
       "3                                     KW-0963~Cytoplasm,\n",
       "4        KW-0256~Endoplasmic reticulum,KW-0472~Membrane,\n",
       "                             ...                        \n",
       "971                   KW-0539~Nucleus,KW-0963~Cytoplasm,\n",
       "972                    KW-0472~Membrane,KW-0539~Nucleus,\n",
       "973                                     KW-0539~Nucleus,\n",
       "974                                     KW-0539~Nucleus,\n",
       "975    KW-0137~Centromere,KW-0158~Chromosome,KW-0206~...\n",
       "Name: UP_KW_CELLULAR_COMPONENT, Length: 976, dtype: object"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc['UP_KW_CELLULAR_COMPONENT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7397c131",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['',\n",
       "       'mrna processing,rna splicing,regulation of mrna export from nucleus,positive regulation of transcription from rna polymerase ii promoter,rna binding,dead/h-box rna helicase binding,',\n",
       "       'rna binding,nucleolus,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,nucleolus,',\n",
       "       'maternal placenta development,regulation of mrna stability,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'rna binding,translation factor activity, rna binding,',\n",
       "       'golgi cisterna membrane,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter in response to stress,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna binding,rna polymerase binding,',\n",
       "       'external side of plasma membrane,',\n",
       "       'nuclear-transcribed mrna catabolic process, nonsense-mediated decay,mrna splicing, via spliceosome,mrna processing,mrna export from nucleus,intracellular mrna localization,rna splicing,mrna transport,regulation of nuclear-transcribed mrna catabolic process, nonsense-mediated decay,rna binding,mrna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase i promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase iii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,rna polymerase i regulatory region sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,nucleolus,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna binding,',\n",
       "       'nuclear-transcribed mrna poly(a) tail shortening,rna binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,external side of plasma membrane,',\n",
       "       'rna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'rrna processing,rna binding,rna helicase activity,',\n",
       "       'rna binding,rna helicase activity,', 'nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'positive regulation of single stranded viral rna replication via double stranded dna intermediate,positive regulation of transcription from rna polymerase ii promoter,rna binding,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii basal transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,negative regulation of transcription from rna polymerase ii promoter in response to stress,nucleolus,',\n",
       "       'positive regulation of transcription elongation from rna polymerase ii promoter,negative regulation of transcription elongation from rna polymerase ii promoter,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'rrna processing,rna binding,nucleolus,',\n",
       "       \"mrna stabilization,negative regulation of gene silencing by mirna,3'-utr-mediated mrna stabilization,rna binding,double-stranded rna binding,mrna binding,mrna 3'-utr binding,mirna binding,mrna 3'-utr au-rich region binding,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'mitochondrial rna processing,mrna processing,regulation of mitochondrial mrna stability,rna binding,rrna binding,',\n",
       "       'rna binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii core promoter sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'rna binding,rna helicase activity,mrna binding,dna/rna helicase activity,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,nucleolus,',\n",
       "       'maturation of ssu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),maturation of ssu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),maturation of ssu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),maturation of ssu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),maturation of ssu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),rrna processing,positive regulation of transcription from rna polymerase i promoter,positive regulation of rrna processing,rna binding,snorna binding,nucleolus,',\n",
       "       'mrna splicing, via spliceosome,regulation of transcription from rna polymerase ii promoter,rna binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter involved in cellular response to chemical stimulus,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of transcription from rna polymerase ii promoter in response to endoplasmic reticulum stress,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'mrna splice site selection,rna binding,',\n",
       "       'trna processing,positive regulation of transcription from rna polymerase ii promoter,trna threonylcarbamoyladenosine metabolic process,',\n",
       "       'negative regulation of receptor internalization,',\n",
       "       'mrna splicing, via spliceosome,mrna processing,mrna catabolic process,rna splicing,rna binding,',\n",
       "       'mrna splicing, via spliceosome,mrna catabolic process,trna processing,rna splicing,maturation of ssu-rrna,rna binding,nucleolus,small nucleolar ribonucleoprotein complex,',\n",
       "       \"regulation of mrna stability,3'-utr-mediated mrna stabilization,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription repressor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,mrna binding,nucleolus,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,negative regulation of transcription from rna polymerase ii promoter in response to endoplasmic reticulum stress,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to stress,positive regulation of transcription from rna polymerase ii promoter in response to oxidative stress,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to hypoxia,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'response to exogenous dsrna,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       \"negative regulation of transcription from rna polymerase ii promoter,transcription elongation from rna polymerase ii promoter,mrna polyadenylation,positive regulation of mrna 3'-end processing,rna polymerase ii core binding,\",\n",
       "       \"mrna 5'-splice site recognition,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna binding,\",\n",
       "       'rna polymerase i transcriptional preinitiation complex assembly,rrna processing,snorna localization,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,rna polymerase i core element sequence-specific dna binding,nucleolus,',\n",
       "       \"trna 5'-leader removal,rrna processing,rna phosphodiester bond hydrolysis, endonucleolytic,ribonuclease p rna binding,rna phosphodiester bond hydrolysis, endonucleolytic,nucleolus,\",\n",
       "       'rrna processing,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,nucleolus,',\n",
       "       'receptor internalization,',\n",
       "       'cellular response to exogenous dsrna,',\n",
       "       'regulation of rna metabolic process,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,',\n",
       "       'regulation of alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,mrna splicing, via spliceosome,rna methylation,mrna processing,mrna export from nucleus,rna splicing,rna binding,mrna binding,',\n",
       "       'mrna splicing, via spliceosome,rna processing,rna binding,',\n",
       "       \"nuclear-transcribed mrna catabolic process, nonsense-mediated decay,regulation of alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,mrna splicing, via spliceosome,rna splicing,negative regulation of mrna splicing, via spliceosome,rna binding,mrna 3'-utr binding,\",\n",
       "       '7-methylguanosine mrna capping,mrna cap binding complex,mrna cap methyltransferase complex,rna binding,mrna (guanine-n7-)-methyltransferase activity,',\n",
       "       \"transcription from rna polymerase i promoter,transcription from rna polymerase iii promoter,dna-directed rna polymerase iii complex,dna-directed rna polymerase i complex,rna polymerase i activity,rna polymerase iii activity,dna-directed 5'-3' rna polymerase activity,\",\n",
       "       \"maintenance of transcriptional fidelity during dna-templated transcription elongation from rna polymerase ii promoter,transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,mrna cleavage,dna-directed rna polymerase ii, core complex,dna-directed 5'-3' rna polymerase activity,nucleolus,\",\n",
       "       \"regulation of transcription from rna polymerase i promoter,transcription from rna polymerase ii promoter,transcription from rna polymerase iii promoter,dna-directed rna polymerase ii, core complex,dna-directed rna polymerase iii complex,dna-directed rna polymerase i complex,dna-directed 5'-3' rna polymerase activity,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of telomerase rna localization to cajal body,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor activity, sequence-specific dna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,primary mirna processing,positive regulation of transcription from rna polymerase ii promoter,regulation of pri-mirna transcription from rna polymerase ii promoter,regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,bacterial-type rna polymerase transcriptional repressor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,dead/h-box rna helicase binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,mirna binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,mirna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rrna processing,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna binding,nucleolus,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase i transcriptional preinitiation complex assembly,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,positive regulation of transcription of nuclear large rrna transcript from rna polymerase i promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase i core element sequence-specific dna binding,rna binding,nucleolus,',\n",
       "       \"mitochondrial mrna catabolic process,positive regulation of mitochondrial rna catabolic process,mitochondrial rna 3'-end processing,rna catabolic process,mitochondrial ncrna surveillance,mitochondrial mrna surveillance,mitochondrial rna surveillance,rna binding,rna helicase activity,double-stranded rna binding,3'-5' rna helicase activity,\",\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter involved in myocardial precursor cell differentiation,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,trna methylation,rna binding,trna (guanine) methyltransferase activity,',\n",
       "       'transcription from rna polymerase ii promoter,rna polymerase ii core promoter sequence-specific dna binding,',\n",
       "       'transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,transcription from rna polymerase iii promoter,mrna transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,rna polymerase ii transcriptional preinitiation complex assembly,positive regulation of transcription initiation from rna polymerase ii promoter,rna polymerase transcription factor sl1 complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii core promoter sequence-specific dna binding,transcription factor activity, core rna polymerase iii binding,rna polymerase ii basal transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'response to exogenous dsrna,',\n",
       "       'modulation by host of viral rna genome replication,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter involved in unfolded protein response,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to endoplasmic reticulum stress,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,',\n",
       "       'regulation of mrna stability,mrna destabilization,rna binding,mrna binding,n6-methyladenosine-containing rna binding,',\n",
       "       \"negative regulation of transcription from rna polymerase ii promoter,nuclear-transcribed mrna catabolic process, deadenylation-dependent decay,nuclear-transcribed mrna poly(a) tail shortening,mrna catabolic process,nuclear-transcribed mrna catabolic process, deadenylation-independent decay,mirna mediated inhibition of translation,regulation of mrna stability,mrna transport,positive regulation of nuclear-transcribed mrna poly(a) tail shortening,3'-utr-mediated mrna destabilization,3'-utr-mediated mrna stabilization,positive regulation of nuclear-transcribed mrna catabolic process, deadenylation-dependent decay,positive regulation of deadenylation-independent decapping of nuclear-transcribed mrna,positive regulation of intracellular mrna localization,positive regulation of gene silencing by mirna,rna binding,mrna binding,mrna 3'-utr au-rich region binding,rna polymerase binding,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to endoplasmic reticulum stress,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'rna processing,trna processing,rna binding,trna-specific adenosine deaminase activity,',\n",
       "       'transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'trna modification,trna aminoacylation for protein translation,alanyl-trna aminoacylation,trna processing,trna binding,aminoacyl-trna editing activity,ser-trna(ala) hydrolase activity,alanine-trna ligase activity,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,mrna polyadenylation,positive regulation of transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of g-protein coupled receptor internalization,rna polymerase ii core promoter proximal region sequence-specific dna binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription elongation from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of production of mirnas involved in gene silencing by mirna,positive regulation of production of mirnas involved in gene silencing by mirna,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,positive regulation of rna polymerase ii transcriptional preinitiation complex assembly,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'rna processing,rna binding,', 'viral rna genome replication,',\n",
       "       'mrna transport,nucleolus,',\n",
       "       'regulation of transcription from rna polymerase ii promoter in response to uv-induced dna damage,',\n",
       "       'receptor internalization,rna binding,double-stranded rna binding,',\n",
       "       'pre-mrna cleavage required for polyadenylation,mrna cleavage and polyadenylation specificity factor complex,rna binding,',\n",
       "       \"positive regulation of mrna splicing, via spliceosome,mrna stabilization,rna binding,mrna 3'-utr binding,small ribosomal subunit rrna binding,\",\n",
       "       'complement activation, alternative pathway,',\n",
       "       \"mrna 3'-utr binding,\",\n",
       "       \"positive regulation of mrna 3'-end processing,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,phosphorylation of rna polymerase ii c-terminal domain,',\n",
       "       'rna polymerase ii carboxy-terminal domain kinase activity,',\n",
       "       'regulation of transcription initiation from rna polymerase ii promoter,rna polymerase ii carboxy-terminal domain kinase activity,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase ii carboxy-terminal domain kinase activity,',\n",
       "       'transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,snrna transcription from rna polymerase ii promoter,snrna transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,phosphorylation of rna polymerase ii c-terminal domain,rna polymerase ii carboxy-terminal domain kinase activity,',\n",
       "       'rrna processing,positive regulation of transcription from rna polymerase ii promoter,rna binding,nucleolus,',\n",
       "       'rna polymerase ii sequence-specific dna binding transcription factor binding,nucleolus,',\n",
       "       'rna processing,rna binding,nucleolus,',\n",
       "       'rna processing,rna metabolic process,rna binding,',\n",
       "       'g-protein coupled receptor internalization,receptor internalization,postsynaptic neurotransmitter receptor internalization,rna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter in response to hypoxia,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'mrna processing,rna splicing,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       \"trna catabolic process,trna 3'-end processing,mitochondrial trna 3'-trailer cleavage, endonucleolytic,mitochondrial trna processing,rna binding,trna-specific ribonuclease activity,3'-trna processing endoribonuclease activity,mitochondrial trna 3'-trailer cleavage, endonucleolytic,\",\n",
       "       'trna wobble uridine modification,trna wobble base 5-methoxycarbonylmethyl-2-thiouridine biosynthesis.,trna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,maternal process involved in parturition,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,negative regulation of transcription elongation from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii core promoter sequence-specific dna binding,primary mirna binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,positive regulation of production of mirnas involved in gene silencing by mirna,positive regulation of production of mirnas involved in gene silencing by mirna,multivesicular body, internal vesicle lumen,',\n",
       "       'positive regulation of receptor internalization,',\n",
       "       'positive regulation of transcription from rna polymerase i promoter,rna polymerase i core binding,',\n",
       "       'mirna mediated inhibition of translation,regulation of gene silencing by mirna,positive regulation of mrna cap binding,positive regulation of mirna mediated inhibition of translation,rna binding,mrna binding,translation factor activity, rna binding,',\n",
       "       \"maturation of 5.8s rrna,nuclear-transcribed mrna catabolic process,rrna processing,rna processing,rna catabolic process,rrna catabolic process,nuclear-transcribed mrna catabolic process, exonucleolytic, 3'-5',u4 snrna 3'-end processing,nuclear mrna surveillance,histone mrna catabolic process,polyadenylation-dependent snorna 3'-end processing,rna phosphodiester bond hydrolysis, exonucleolytic,nuclear exosome (rnase complex),cytoplasmic exosome (rnase complex),exosome (rnase complex),nucleolar exosome (rnase complex),mrna 3'-utr au-rich region binding,nuclear-transcribed mrna catabolic process, exonucleolytic, 3'-5',rna phosphodiester bond hydrolysis, exonucleolytic,nucleolus,nucleolar exosome (rnase complex),\",\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription repressor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,mitochondrial rna polymerase binding promoter specificity activity,',\n",
       "       'transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcriptional preinitiation complex assembly,positive regulation of transcription initiation from rna polymerase ii promoter,rna polymerase ii basal transcription factor binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'transcription from rna polymerase ii promoter,transcription initiation from rna polymerase ii promoter,rna binding,',\n",
       "       'trna aminoacylation for protein translation,glutamyl-trna aminoacylation,prolyl-trna aminoacylation,aminoacyl-trna synthetase multienzyme complex,glutamate-trna ligase activity,proline-trna ligase activity,rna stem-loop binding,',\n",
       "       'receptor internalization,external side of plasma membrane,',\n",
       "       'mrna catabolic process,positive regulation of rna splicing,negative regulation of transcription from rna polymerase ii promoter in response to stress,rna binding,',\n",
       "       'mrna splicing, via spliceosome,mrna processing,rna splicing,positive regulation of mrna splicing, via spliceosome,rna binding,nucleolus,',\n",
       "       'golgi cisterna,rna binding,double-stranded rna binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter in response to iron,regulation of transcription from rna polymerase ii promoter in response to oxidative stress,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,negative regulation of single stranded viral rna replication via double stranded dna intermediate,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'internal protein amino acid acetylation,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,',\n",
       "       \"mitochondrial trna methylation,mitochondrial trna 5'-end processing,mitochondrial trna 3'-end processing,trna methyltransferase complex,trna binding,rna binding,\",\n",
       "       'regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter in response to oxidative stress,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to hypoxia,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,bacterial-type rna polymerase transcriptional activator activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       \"rna secondary structure unwinding,trna binding,rna binding,single-stranded rna binding,rna-dependent atpase activity,5'-3' rna helicase activity,\",\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'mrna stabilization,',\n",
       "       \"mrna transport,regulation of rna metabolic process,crd-mediated mrna stabilization,rna binding,mrna binding,mrna 3'-utr binding,mrna 5'-utr binding,n6-methyladenosine-containing rna binding,\",\n",
       "       'insulin-like growth factor ternary complex,',\n",
       "       'snrna processing,regulation of transcription elongation from rna polymerase ii promoter,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,positive regulation of rna biosynthetic process,',\n",
       "       'trna aminoacylation for protein translation,isoleucyl-trna aminoacylation,trna binding,aminoacyl-trna editing activity,isoleucine-trna ligase activity,',\n",
       "       'mrna processing,rna splicing,oxidative rna demethylation,positive regulation of transcription from rna polymerase ii promoter,regulation of mrna splicing, via spliceosome,rna binding,single-stranded rna binding,oxidative rna demethylase activity,nucleolus,',\n",
       "       'rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'mrna transport,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii sequence-specific dna binding transcription factor binding,nucleolus,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,',\n",
       "       'rna binding,mrna binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter in response to stress,regulation of transcription from rna polymerase ii promoter in response to stress,rna polymerase ii core promoter proximal region sequence-specific dna binding,',\n",
       "       'regulation of alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,mrna processing,rna splicing,regulation of rna splicing,regulatory region rna binding,rna binding,double-stranded rna binding,',\n",
       "       'regulation of alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,mrna processing,rna splicing,regulation of rna splicing,rna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'rna binding,rrna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter involved in myocardial precursor cell differentiation,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to hypoxia,',\n",
       "       'rrna processing,rna binding,regulation of protein localization to nucleolus,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,maternal behavior,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,',\n",
       "       'rrna processing,box h/aca snorna metabolic process,rna binding,box c/d snorna binding,box h/aca snorna binding,nucleolus organization,nucleolus,',\n",
       "       'mrna splice site selection,rna splicing,rna binding,nucleolus,',\n",
       "       'transcription-dependent tethering of rna polymerase ii gene dna at nuclear periphery,mrna export from nucleus,poly(a)+ mrna export from nucleus,',\n",
       "       'rna export from nucleus,mrna transport,',\n",
       "       'mrna export from nucleus,mrna transport,',\n",
       "       'poly(a)+ mrna export from nucleus,mrna transport,',\n",
       "       'mrna splicing, via spliceosome,rna binding,mrna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,negative regulation of gene silencing by mirna,negative regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'response to dsrna,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,nucleolus,',\n",
       "       'nuclear-transcribed mrna poly(a) tail shortening,mrna processing,positive regulation of cytoplasmic mrna processing body assembly,rna phosphodiester bond hydrolysis, exonucleolytic,rna phosphodiester bond hydrolysis, exonucleolytic,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,nucleolus,',\n",
       "       'alternative mrna splicing, via spliceosome,alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,regulation of alternative mrna splicing, via spliceosome,mrna splice site selection,mrna processing,rna splicing,rna binding,',\n",
       "       'rna splicing, via transesterification reactions,mrna splicing, via spliceosome,rna processing,rna splicing,u6 snrna binding,u4 snrna binding,',\n",
       "       'regulation of mrna stability,', 'mrna processing,',\n",
       "       'maternal process involved in female pregnancy,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'terminal cisterna,',\n",
       "       \"trna 5'-leader removal,rrna processing,rna phosphodiester bond hydrolysis, endonucleolytic,ribonuclease p rna binding,rna phosphodiester bond hydrolysis, endonucleolytic,nucleolar ribonuclease p complex,nucleolus,\",\n",
       "       'mrna catabolic process,',\n",
       "       'transcription-dependent tethering of rna polymerase ii gene dna at nuclear periphery,rna export from nucleus,mrna export from nucleus,rna binding,',\n",
       "       'rrna processing,mrna processing,rna splicing,regulation of rna splicing,positive regulation of transcription from rna polymerase ii promoter,rna binding,nucleolus,',\n",
       "       'rna binding,mrna binding,rrna binding,',\n",
       "       'endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),rna binding,5s rrna binding,endonucleolytic cleavage in its1 to separate ssu-rrna from 5.8s rrna and lsu-rrna from tricistronic rrna transcript (ssu-rrna, 5.8s rrna, lsu-rrna),protein localization to nucleolus,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii core promoter sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,nucleolus,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,mrna transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of mirna mediated inhibition of translation,negative regulation of primary mirna processing,positive regulation of gene silencing by mirna,rna polymerase ii transcription factor complex,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,rna polymerase ii transcription factor activity, ligand-activated sequence-specific dna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,primary mirna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,mrna splicing, via spliceosome,regulation of transcription from rna polymerase ii promoter,rna binding,mrna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,',\n",
       "       'positive regulation of receptor internalization,receptor internalization,',\n",
       "       'rna splicing, via transesterification reactions,mrna processing,rna binding,',\n",
       "       'neurotransmitter receptor internalization,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,receptor internalization,',\n",
       "       'trna-type intron splice site recognition and cleavage,trna splicing, via endonucleolytic cleavage and ligation,mrna processing,rna phosphodiester bond hydrolysis, endonucleolytic,trna-intron endonuclease complex,trna-intron endonuclease activity,trna splicing, via endonucleolytic cleavage and ligation,rna phosphodiester bond hydrolysis, endonucleolytic,nucleolus,',\n",
       "       \"transcription, rna-templated,rna-dependent dna biosynthetic process,production of sirna involved in rna interference,production of sirna involved in rna interference,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of production of sirna involved in rna interference,negative regulation of production of sirna involved in rna interference,negative regulation of production of mirnas involved in gene silencing by mirna,negative regulation of production of mirnas involved in gene silencing by mirna,rna-directed rna polymerase complex,rna-directed rna polymerase complex,trna binding,telomerase rna reverse transcriptase activity,rna binding,rna-directed dna polymerase activity,rna-directed 5'-3' rna polymerase activity,rna-directed 5'-3' rna polymerase activity,telomerase rna binding,template-free rna nucleotidyltransferase,positive regulation of protein localization to nucleolus,nucleolus,\",\n",
       "       \"mrna processing,rna 3'-end processing,rna polyadenylation,negative regulation of nuclear-transcribed mrna poly(a) tail shortening,rna 3' uridylation,positive regulation of 3'-utr-mediated mrna stabilization,nucleolus,\",\n",
       "       'mrna splicing, via spliceosome,',\n",
       "       'transcription elongation from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,mrna processing,rna splicing,positive regulation of transcription elongation from rna polymerase ii promoter,negative regulation of transcription elongation from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna binding,rna polymerase ii sequence-specific dna binding transcription factor binding,rna polymerase binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii transcription factor complex,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'mirna transport,external side of plasma membrane,',\n",
       "       \"negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,mrna transcription,rrna transcription,positive regulation of transcription from rna polymerase ii promoter in response to stress,positive regulation of rna polymerase ii transcriptional preinitiation complex assembly,positive regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter in response to hypoxia,positive regulation of pri-mirna transcription from rna polymerase ii promoter,positive regulation of pri-mirna transcription from rna polymerase ii promoter,negative regulation of production of mirnas involved in gene silencing by mirna,negative regulation of production of mirnas involved in gene silencing by mirna,positive regulation of production of mirnas involved in gene silencing by mirna,positive regulation of production of mirnas involved in gene silencing by mirna,regulation of transcription from rna polymerase ii promoter in response to dna damage,positive regulation of transcription from rna polymerase ii promoter in response to endoplasmic reticulum stress,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,bacterial-type rna polymerase transcriptional activator activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,mrna 3'-utr binding,rna polymerase ii sequence-specific dna binding transcription factor binding,nucleolus,\",\n",
       "       'negative regulation of g-protein coupled receptor internalization,',\n",
       "       'maternal process involved in female pregnancy,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,regulation of rna splicing,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,nucleolus,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,positive regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,transcriptional activator activity, rna polymerase ii transcription regulatory region sequence-specific binding,',\n",
       "       'positive regulation of transcription from rna polymerase ii promoter,negative regulation of transcription initiation from rna polymerase ii promoter,',\n",
       "       'regulation of transcription from rna polymerase ii promoter,rna polymerase ii core promoter proximal region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,',\n",
       "       'negative regulation of transcription from rna polymerase ii promoter,regulation of transcription from rna polymerase ii promoter,rna polymerase ii regulatory region sequence-specific dna binding,rna polymerase ii transcription factor activity, sequence-specific dna binding,transcriptional repressor activity, rna polymerase ii transcription regulatory region sequence-specific binding,'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene_cats_bpcc['UP_SEQ_FEATURE']\n",
    "gene_cats_bpcc['RNA_nucleoli'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "cbca302a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "top_channel\n",
       "max_AGP     238\n",
       "max_DNA     227\n",
       "max_ER      182\n",
       "max_Mito    176\n",
       "max_RNA     155\n",
       "dtype: int64"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dfcats\n",
    "# gene_cats_bpcc['UP_SEQ_FEATURE'].str.lower()\n",
    "# gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).str.lower()\n",
    "# gene_cats_bpcc\n",
    "\n",
    "# gene_cats_bpcc['UP_KW_CELLULAR_COMPONENT'].unique()\n",
    "# ci=2\n",
    "# Channelss_cats=['DNA|Nuclei_AreaShape','RNA','AGP|Cytoplasm_AreaShape|Cells_AreaShape','Mito','ER']\n",
    "# dfcats.loc[:,dfcats.columns.str.contains(Channelss_cats[ci])]\n",
    "# dfcats[['max_'+Channelss[ci] for ci in range(len(Channelss))]].idxmax(axis=\"columns\")\n",
    "dfcats['top_channel']=dfcats[['max_'+Channelss[ci] for ci in range(len(Channelss))]].idxmax(axis=\"columns\")\n",
    "\n",
    "dfcats.groupby('top_channel').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "b54d546a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['' 'steroid metabolic process,'\n",
      " 'steroid hormone receptor complex assembly,go:0006825~copper ion transport,go:0007566~embryo implantation,go:0010977~negative regulation of neuron projection development,go:0030521~androgen receptor signaling pathway,go:0030850~prostate gland development,go:0031111~negative regulation of microtubule polymerization or depolymerization,go:0031115~negative regulation of microtubule polymerization,go:0031503~protein complex localization,go:0046661~male sex differentiation,go:0061077~chaperone-mediated protein folding,'\n",
      " 'intracellular steroid hormone receptor signaling pathway,go:0032024~positive regulation of insulin secretion,go:0032962~positive regulation of inositol trisphosphate biosynthetic process,go:0042311~vasodilation,go:0043065~positive regulation of apoptotic process,go:0043280~positive regulation of cysteine-type endopeptidase activity involved in apoptotic process,go:0043401~steroid hormone mediated signaling pathway,go:0043410~positive regulation of mapk cascade,go:0045087~innate immune response,go:0045599~negative regulation of fat cell differentiation,go:0045742~positive regulation of epidermal growth factor receptor signaling pathway,go:0045745~positive regulation of g-protein coupled receptor protein signaling pathway,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0050728~negative regulation of inflammatory response,go:0050769~positive regulation of neurogenesis,go:0050804~modulation of synaptic transmission,go:0051055~negative regulation of lipid biosynthetic process,go:0051281~positive regulation of release of sequestered calcium ion into cytosol,go:0051480~regulation of cytosolic calcium ion concentration,go:0051726~regulation of cell cycle,go:0051898~negative regulation of protein kinase b signaling,go:0070373~negative regulation of erk1 and erk2 cascade,go:0070374~positive regulation of erk1 and erk2 cascade,go:0070474~positive regulation of uterine smooth muscle contraction,go:0071333~cellular response to glucose stimulus,go:0071356~cellular response to tumor necrosis factor,go:0071375~cellular response to peptide hormone stimulus,go:0071389~cellular response to mineralocorticoid stimulus,go:0071392~cellular response to estradiol stimulus,go:0090200~positive regulation of release of cytochrome c from mitochondria,go:1903078~positive regulation of protein localization to plasma membrane,go:1904706~negative regulation of vascular smooth muscle cell proliferation,go:2000353~positive regulation of endothelial cell apoptotic process,go:2000724~positive regulation of cardiac vascular smooth muscle cell differentiation,go:2001238~positive regulation of extrinsic apoptotic signaling pathway,steroid hormone mediated signaling pathway,go:0043410~positive regulation of mapk cascade,go:0045087~innate immune response,go:0045599~negative regulation of fat cell differentiation,go:0045742~positive regulation of epidermal growth factor receptor signaling pathway,go:0045745~positive regulation of g-protein coupled receptor protein signaling pathway,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0050728~negative regulation of inflammatory response,go:0050769~positive regulation of neurogenesis,go:0050804~modulation of synaptic transmission,go:0051055~negative regulation of lipid biosynthetic process,go:0051281~positive regulation of release of sequestered calcium ion into cytosol,go:0051480~regulation of cytosolic calcium ion concentration,go:0051726~regulation of cell cycle,go:0051898~negative regulation of protein kinase b signaling,go:0070373~negative regulation of erk1 and erk2 cascade,go:0070374~positive regulation of erk1 and erk2 cascade,go:0070474~positive regulation of uterine smooth muscle contraction,go:0071333~cellular response to glucose stimulus,go:0071356~cellular response to tumor necrosis factor,go:0071375~cellular response to peptide hormone stimulus,go:0071389~cellular response to mineralocorticoid stimulus,go:0071392~cellular response to estradiol stimulus,go:0090200~positive regulation of release of cytochrome c from mitochondria,go:1903078~positive regulation of protein localization to plasma membrane,go:1904706~negative regulation of vascular smooth muscle cell proliferation,go:2000353~positive regulation of endothelial cell apoptotic process,go:2000724~positive regulation of cardiac vascular smooth muscle cell differentiation,go:2001238~positive regulation of extrinsic apoptotic signaling pathway,'\n",
      " 'regulation of steroid biosynthetic process,go:0071356~cellular response to tumor necrosis factor,go:1903507~negative regulation of nucleic acid-templated transcription,'\n",
      " 'steroid biosynthetic process,go:0006695~cholesterol biosynthetic process,go:0007224~smoothened signaling pathway,go:0008203~cholesterol metabolic process,go:0060716~labyrinthine layer blood vessel development,'\n",
      " 'cellular response to steroid hormone stimulus,go:0071404~cellular response to low-density lipoprotein particle stimulus,go:0090150~establishment of protein localization to membrane,'\n",
      " 'steroid biosynthetic process,'\n",
      " 'steroid hormone mediated signaling pathway,go:0043407~negative regulation of map kinase activity,go:0043410~positive regulation of mapk cascade,go:0045603~positive regulation of endothelial cell differentiation,go:0045606~positive regulation of epidermal cell differentiation,go:0045662~negative regulation of myoblast differentiation,go:0045666~positive regulation of neuron differentiation,go:0045669~positive regulation of osteoblast differentiation,go:0045778~positive regulation of ossification,go:0045786~negative regulation of cell cycle,go:0045839~negative regulation of mitotic nuclear division,go:0045843~negative regulation of striated muscle tissue development,go:0045892~negative regulation of transcription, dna-templated,go:0045893~positive regulation of transcription, dna-templated,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0048286~lung alveolus development,go:0048392~intermediate mesodermal cell differentiation,go:0048660~regulation of smooth muscle cell proliferation,go:0048661~positive regulation of smooth muscle cell proliferation,go:0048663~neuron fate commitment,go:0048701~embryonic cranial skeleton morphogenesis,go:0048745~smooth muscle tissue development,go:0048754~branching morphogenesis of an epithelial tube,go:0050679~positive regulation of epithelial cell proliferation,go:0050680~negative regulation of epithelial cell proliferation,go:0050918~positive chemotaxis,go:0051150~regulation of smooth muscle cell differentiation,go:0055007~cardiac muscle cell differentiation,go:0055020~positive regulation of cardiac muscle fiber development,go:0060113~inner ear receptor cell differentiation,go:0060197~cloacal septation,go:0060235~lens induction in camera-type eye,go:0060272~embryonic skeletal joint morphogenesis,go:0060363~cranial suture morphogenesis,go:0060391~positive regulation of smad protein import into nucleus,go:0060393~regulation of pathway-restricted smad protein phosphorylation,go:0060395~smad protein signal transduction,go:0060425~lung morphogenesis,go:0060433~bronchus development,go:0060438~trachea development,go:0060440~trachea formation,go:0060441~epithelial tube branching involved in lung morphogenesis,go:0060442~branching involved in prostate gland morphogenesis,go:0060449~bud elongation involved in lung branching,go:0060502~epithelial cell proliferation involved in lung morphogenesis,go:0060503~bud dilation involved in lung branching,go:0060513~prostatic bud formation,go:0060548~negative regulation of cell death,go:0060592~mammary gland formation,go:0060684~epithelial-mesenchymal cell signaling,go:0060686~negative regulation of prostatic bud formation,go:0060687~regulation of branching involved in prostate gland morphogenesis,go:0060976~coronary vasculature development,go:0061036~positive regulation of cartilage development,go:0061047~positive regulation of branching involved in lung morphogenesis,go:0061149~bmp signaling pathway involved in ureter morphogenesis,go:0061151~bmp signaling pathway involved in renal system segmentation,go:0061155~pulmonary artery endothelial tube morphogenesis,go:0061312~bmp signaling pathway involved in heart development,go:0061626~pharyngeal arch artery morphogenesis,go:0070244~negative regulation of thymocyte apoptotic process,go:0070374~positive regulation of erk1 and erk2 cascade,go:0071773~cellular response to bmp stimulus,go:0071893~bmp signaling pathway involved in nephric duct formation,go:0072015~glomerular visceral epithelial cell development,go:0072097~negative regulation of branch elongation involved in ureteric bud branching by bmp signaling pathway,go:0072101~specification of ureteric bud anterior/posterior symmetry by bmp signaling pathway,go:0072104~glomerular capillary formation,go:0072125~negative regulation of glomerular mesangial cell proliferation,go:0072138~mesenchymal cell proliferation involved in ureteric bud development,go:0072161~mesenchymal cell differentiation involved in kidney development,go:0072192~ureter epithelial cell differentiation,go:0072193~ureter smooth muscle cell differentiation,go:0072198~mesenchymal cell proliferation involved in ureter development,go:0072200~negative regulation of mesenchymal cell proliferation involved in ureter development,go:0072205~metanephric collecting duct development,go:0090184~positive regulation of kidney development,go:0090191~negative regulation of branching involved in ureteric bud morphogenesis,go:0090194~negative regulation of glomerulus development,go:1900745~positive regulation of p38mapk cascade,go:1901964~positive regulation of cell proliferation involved in outflow tract morphogenesis,go:1902894~negative regulation of pri-mirna transcription from rna polymerase ii promoter,go:1902895~positive regulation of pri-mirna transcription from rna polymerase ii promoter,go:1903800~positive regulation of production of mirnas involved in gene silencing by mirna,go:1904238~pericyte cell differentiation,go:1905312~positive regulation of cardiac neural crest cell migration involved in outflow tract morphogenesis,go:1905770~regulation of mesodermal cell differentiation,go:2000005~negative regulation of metanephric s-shaped body morphogenesis,go:2000007~negative regulation of metanephric comma-shaped body morphogenesis,go:2000137~negative regulation of cell proliferation involved in heart morphogenesis,go:2001237~negative regulation of extrinsic apoptotic signaling pathway,'\n",
      " 'response to corticosteroid,go:0032355~response to estradiol,go:0032964~collagen biosynthetic process,go:0034504~protein localization to nucleus,go:0034505~tooth mineralization,go:0038063~collagen-activated tyrosine kinase receptor signaling pathway,go:0042060~wound healing,go:0042493~response to drug,go:0042542~response to hydrogen peroxide,go:0043434~response to peptide hormone,go:0043588~skin development,go:0043589~skin morphogenesis,go:0044344~cellular response to fibroblast growth factor stimulus,go:0044691~tooth eruption,go:0045893~positive regulation of transcription, dna-templated,go:0048706~embryonic skeletal system development,go:0051591~response to camp,go:0055093~response to hyperoxia,go:0060325~face morphogenesis,go:0060346~bone trabecula formation,go:0060351~cartilage development involved in endochondral bone morphogenesis,go:0071230~cellular response to amino acid stimulus,go:0071260~cellular response to mechanical stimulus,go:0071300~cellular response to retinoic acid,go:0071306~cellular response to vitamin e,go:0071356~cellular response to tumor necrosis factor,go:0071364~cellular response to epidermal growth factor stimulus,go:0071560~cellular response to transforming growth factor beta stimulus,go:0090263~positive regulation of canonical wnt signaling pathway,go:1902618~cellular response to fluoride,'\n",
      " 'steroid biosynthetic process,go:0006695~cholesterol biosynthetic process,go:0008610~lipid biosynthetic process,go:0009058~biosynthetic process,go:0016126~sterol biosynthetic process,go:0045338~farnesyl diphosphate metabolic process,'\n",
      " 'cellular response to steroid hormone stimulus,go:0090063~positive regulation of microtubule nucleation,go:0090084~negative regulation of inclusion body assembly,go:0097201~negative regulation of transcription from rna polymerase ii promoter in response to stress,go:1901029~negative regulation of mitochondrial outer membrane permeabilization involved in apoptotic signaling pathway,go:1901673~regulation of mitotic spindle assembly,go:1902236~negative regulation of endoplasmic reticulum stress-induced intrinsic apoptotic signaling pathway,go:1902380~positive regulation of endoribonuclease activity,go:1903265~positive regulation of tumor necrosis factor-mediated signaling pathway,go:2001240~negative regulation of extrinsic apoptotic signaling pathway in absence of ligand,'\n",
      " 'cellular response to steroid hormone stimulus,go:0072318~clathrin coat disassembly,go:1902904~negative regulation of supramolecular fiber organization,go:1904589~regulation of protein import,go:1904764~chaperone-mediated autophagy translocation complex disassembly,go:1990832~slow axonal transport,'\n",
      " 'negative regulation of intracellular steroid hormone receptor signaling pathway,go:0035556~intracellular signal transduction,go:0035978~histone h2a-s139 phosphorylation,go:0035986~senescence-associated heterochromatin focus assembly,go:0035987~endodermal cell differentiation,go:0035988~chondrocyte proliferation,go:0040008~regulation of growth,go:0042769~dna damage response, detection of dna damage,go:0043065~positive regulation of apoptotic process,go:0043066~negative regulation of apoptotic process,go:0043392~negative regulation of dna binding,go:0043922~negative regulation by host of viral transcription,go:0045444~fat cell differentiation,go:0045766~positive regulation of angiogenesis,go:0045869~negative regulation of single stranded viral rna replication via double stranded dna intermediate,go:0045892~negative regulation of transcription, dna-templated,go:0045893~positive regulation of transcription, dna-templated,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0046426~negative regulation of jak-stat cascade,go:0048333~mesodermal cell differentiation,go:0048762~mesenchymal cell differentiation,go:0048863~stem cell differentiation,go:0051091~positive regulation of sequence-specific dna binding transcription factor activity,go:0051301~cell division,go:0051726~regulation of cell cycle,go:0071864~positive regulation of cell proliferation in bone marrow,go:0071902~positive regulation of protein serine/threonine kinase activity,go:0090402~oncogene-induced cell senescence,go:2000036~regulation of stem cell population maintenance,go:2000648~positive regulation of stem cell proliferation,go:2000685~positive regulation of cellular response to x-ray,go:2000773~negative regulation of cellular senescence,go:2000774~positive regulation of cellular senescence,go:2001022~positive regulation of response to dna damage stimulus,go:2001033~negative regulation of double-strand break repair via nonhomologous end joining,go:2001038~regulation of cellular response to drug,'\n",
      " \"c21-steroid hormone metabolic process,go:0008209~androgen metabolic process,go:0008210~estrogen metabolic process,go:0051289~protein homotetramerization,go:0070901~mitochondrial trna methylation,go:0097745~mitochondrial trna 5'-end processing,go:1990180~mitochondrial trna 3'-end processing,\"\n",
      " 'negative regulation of steroid biosynthetic process,go:0016126~sterol biosynthetic process,go:0032869~cellular response to insulin stimulus,go:0032933~srebp signaling pathway,go:0036315~cellular response to sterol,go:0036316~srebp-scap complex retention in endoplasmic reticulum,go:0042472~inner ear morphogenesis,go:0042474~middle ear morphogenesis,go:0042632~cholesterol homeostasis,go:0045599~negative regulation of fat cell differentiation,go:0045717~negative regulation of fatty acid biosynthetic process,go:0060021~palate development,go:0060363~cranial suture morphogenesis,go:0070862~negative regulation of protein exit from endoplasmic reticulum,go:1901303~negative regulation of cargo loading into copii-coated vesicle,'\n",
      " 'positive regulation of steroid hormone biosynthetic process,go:0090398~cellular senescence,go:0097062~dendritic spine maintenance,go:0097242~beta-amyloid clearance,go:1902065~response to l-glutamate,go:1903944~negative regulation of hepatocyte apoptotic process,go:1904045~cellular response to aldosterone,go:1904193~negative regulation of cholangiocyte apoptotic process,go:1904385~cellular response to angiotensin,go:1904646~cellular response to beta-amyloid,go:1990314~cellular response to insulin-like growth factor stimulus,'\n",
      " 'response to corticosteroid,go:0032495~response to muramyl dipeptide,go:0032496~response to lipopolysaccharide,go:0035116~embryonic hindlimb morphogenesis,go:0035148~tube formation,go:0035914~skeletal muscle cell differentiation,go:0035924~cellular response to vascular endothelial growth factor stimulus,go:0042246~tissue regeneration,go:0042670~retinal cone cell differentiation,go:0043086~negative regulation of catalytic activity,go:0045070~positive regulation of viral genome replication,go:0045603~positive regulation of endothelial cell differentiation,go:0045608~negative regulation of auditory receptor cell differentiation,go:0045618~positive regulation of keratinocyte differentiation,go:0045662~negative regulation of myoblast differentiation,go:0045668~negative regulation of osteoblast differentiation,go:0045747~positive regulation of notch signaling pathway,go:0045892~negative regulation of transcription, dna-templated,go:0045893~positive regulation of transcription, dna-templated,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0045955~negative regulation of calcium ion-dependent exocytosis,go:0045967~negative regulation of growth rate,go:0046427~positive regulation of jak-stat cascade,go:0046533~negative regulation of photoreceptor cell differentiation,go:0046579~positive regulation of ras protein signal transduction,go:0048103~somatic stem cell division,go:0048708~astrocyte differentiation,go:0048709~oligodendrocyte differentiation,go:0048711~positive regulation of astrocyte differentiation,go:0048715~negative regulation of oligodendrocyte differentiation,go:0048754~branching morphogenesis of an epithelial tube,go:0048873~homeostasis of number of cells within a tissue,go:0050673~epithelial cell proliferation,go:0050679~positive regulation of epithelial cell proliferation,go:0050680~negative regulation of epithelial cell proliferation,go:0050768~negative regulation of neurogenesis,go:0051726~regulation of cell cycle,go:0055008~cardiac muscle tissue morphogenesis,go:0060038~cardiac muscle cell proliferation,go:0060045~positive regulation of cardiac muscle cell proliferation,go:0060253~negative regulation of glial cell proliferation,go:0060271~cilium assembly,go:0060317~cardiac epithelial to mesenchymal transition,go:0060354~negative regulation of cell adhesion molecule production,go:0060379~cardiac muscle cell myoblast differentiation,go:0060411~cardiac septum morphogenesis,go:0060412~ventricular septum morphogenesis,go:0060528~secretory columnal luminar epithelial cell differentiation involved in prostate glandular acinus development,go:0060740~prostate gland epithelium morphogenesis,go:0060768~regulation of epithelial cell proliferation involved in prostate gland development,go:0060842~arterial endothelial cell differentiation,go:0060843~venous endothelial cell differentiation,go:0060948~cardiac vascular smooth muscle cell development,go:0060956~endocardial cell differentiation,go:0060979~vasculogenesis involved in coronary vascular morphogenesis,go:0060982~coronary artery morphogenesis,go:0061101~neuroendocrine cell differentiation,go:0061314~notch signaling involved in heart development,go:0061384~heart trabecula morphogenesis,go:0061419~positive regulation of transcription from rna polymerase ii promoter in response to hypoxia,go:0070168~negative regulation of biomineral tissue development,go:0070374~positive regulation of erk1 and erk2 cascade,go:0070986~left/right axis specification,go:0071228~cellular response to tumor cell,go:0071372~cellular response to follicle-stimulating hormone stimulus,go:0072017~distal tubule development,go:0072044~collecting duct development,go:0072091~regulation of stem cell proliferation,go:0072144~glomerular mesangial cell development,go:0072148~epithelial cell fate commitment,go:0072538~t-helper 17 type immune response,go:0090051~negative regulation of cell migration involved in sprouting angiogenesis,go:0090090~negative regulation of canonical wnt signaling pathway,go:0097150~neuronal stem cell population maintenance,go:0097400~interleukin-17-mediated signaling pathway,go:1901201~regulation of extracellular matrix assembly,go:1902263~apoptotic process involved in embryonic digit morphogenesis,go:1902339~positive regulation of apoptotic process involved in morphogenesis,go:1903849~positive regulation of aorta morphogenesis,go:2000048~negative regulation of cell-cell adhesion mediated by cadherin,go:2000737~negative regulation of stem cell differentiation,go:2000811~negative regulation of anoikis,go:2000974~negative regulation of pro-b cell differentiation,go:2001027~negative regulation of endothelial cell chemotaxis,'\n",
      " 'intracellular steroid hormone receptor signaling pathway,go:0031946~regulation of glucocorticoid biosynthetic process,go:0042711~maternal behavior,go:0042921~glucocorticoid receptor signaling pathway,go:0043402~glucocorticoid mediated signaling pathway,go:0043525~positive regulation of neuron apoptotic process,go:0045892~negative regulation of transcription, dna-templated,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0051301~cell division,go:0060603~mammary gland duct morphogenesis,go:0071383~cellular response to steroid hormone stimulus,go:0071385~cellular response to glucocorticoid stimulus,go:0071549~cellular response to dexamethasone stimulus,go:0071560~cellular response to transforming growth factor beta stimulus,go:1902895~positive regulation of pri-mirna transcription from rna polymerase ii promoter,cellular response to steroid hormone stimulus,go:0071385~cellular response to glucocorticoid stimulus,go:0071549~cellular response to dexamethasone stimulus,go:0071560~cellular response to transforming growth factor beta stimulus,go:1902895~positive regulation of pri-mirna transcription from rna polymerase ii promoter,'\n",
      " 'steroid hormone mediated signaling pathway,go:0043415~positive regulation of skeletal muscle tissue regeneration,go:0043616~keratinocyte proliferation,go:0045598~regulation of fat cell differentiation,go:0045600~positive regulation of fat cell differentiation,go:0045662~negative regulation of myoblast differentiation,go:0045684~positive regulation of epidermis development,go:0045892~negative regulation of transcription, dna-templated,go:0045893~positive regulation of transcription, dna-templated,go:0045923~positive regulation of fatty acid metabolic process,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0046321~positive regulation of fatty acid oxidation,go:0046697~decidualization,go:0048662~negative regulation of smooth muscle cell proliferation,go:0050680~negative regulation of epithelial cell proliferation,go:0050728~negative regulation of inflammatory response,go:0051546~keratinocyte migration,go:0060612~adipose tissue development,go:0070341~fat cell proliferation,go:0070346~positive regulation of fat cell proliferation,go:0071222~cellular response to lipopolysaccharide,go:0071456~cellular response to hypoxia,go:0097190~apoptotic signaling pathway,go:1902894~negative regulation of pri-mirna transcription from rna polymerase ii promoter,go:1904659~glucose transmembrane transport,go:2000288~positive regulation of myoblast proliferation,'\n",
      " 'negative regulation of intracellular steroid hormone receptor signaling pathway,go:0033688~regulation of osteoblast proliferation,go:0034329~cell junction assembly,go:0034446~substrate adhesion-dependent cell spreading,go:0035385~roundabout signaling pathway,go:0036089~cleavage furrow formation,go:0038027~apolipoprotein a-i-mediated signaling pathway,go:0042476~odontogenesis,go:0042493~response to drug,go:0043123~positive regulation of i-kappab kinase/nf-kappab signaling,go:0043124~negative regulation of i-kappab kinase/nf-kappab signaling,go:0043149~stress fiber assembly,go:0043200~response to amino acid,go:0043280~positive regulation of cysteine-type endopeptidase activity involved in apoptotic process,go:0043297~apical junction assembly,go:0043366~beta selection,go:0043524~negative regulation of neuron apoptotic process,go:0043525~positive regulation of neuron apoptotic process,go:0043542~endothelial cell migration,go:0043931~ossification involved in bone maturation,go:0044319~wound healing, spreading of cells,go:0045198~establishment of epithelial cell apical/basal polarity,go:0045471~response to ethanol,go:0045665~negative regulation of neuron differentiation,go:0045666~positive regulation of neuron differentiation,go:0045727~positive regulation of translation,go:0045792~negative regulation of cell size,go:0046039~gtp metabolic process,go:0046638~positive regulation of alpha-beta t cell differentiation,go:0048812~neuron projection morphogenesis,go:0050773~regulation of dendrite development,go:0050919~negative chemotaxis,go:0051384~response to glucocorticoid,go:0051496~positive regulation of stress fiber assembly,go:0051893~regulation of focal adhesion assembly,go:0051924~regulation of calcium ion transport,go:0060071~wnt signaling pathway, planar cell polarity pathway,go:0060193~positive regulation of lipase activity,go:0061383~trabecula morphogenesis,go:0070507~regulation of microtubule cytoskeleton organization,go:0071222~cellular response to lipopolysaccharide,go:0071345~cellular response to cytokine stimulus,go:0071803~positive regulation of podosome assembly,go:0071902~positive regulation of protein serine/threonine kinase activity,go:0090051~negative regulation of cell migration involved in sprouting angiogenesis,go:0090307~mitotic spindle assembly,go:0090324~negative regulation of oxidative phosphorylation,go:0097049~motor neuron apoptotic process,go:0097498~endothelial tube lumen extension,go:1901224~positive regulation of nik/nf-kappab signaling,go:1902766~skeletal muscle satellite cell migration,go:1903427~negative regulation of reactive oxygen species biosynthetic process,go:1903673~mitotic cleavage furrow formation,go:1904695~positive regulation of vascular smooth muscle contraction,go:1904706~negative regulation of vascular smooth muscle cell proliferation,go:1904753~negative regulation of vascular associated smooth muscle cell migration,go:1904996~positive regulation of leukocyte adhesion to vascular endothelial cell,go:1905244~regulation of modification of synaptic structure,go:1905274~regulation of modification of postsynaptic actin cytoskeleton,go:1990869~cellular response to chemokine,go:2000177~regulation of neural precursor cell proliferation,go:2000406~positive regulation of t cell migration,go:2000672~negative regulation of motor neuron apoptotic process,'\n",
      " 'response to steroid hormone,go:0048685~negative regulation of collateral sprouting of intact axon in response to injury,go:0071394~cellular response to testosterone stimulus,go:2000866~positive regulation of estradiol secretion,'\n",
      " \"regulation of steroid metabolic process,go:0019221~cytokine-mediated signaling pathway,go:0019530~taurine metabolic process,go:0019915~lipid storage,go:0030183~b cell differentiation,go:0030218~erythrocyte differentiation,go:0030856~regulation of epithelial cell differentiation,go:0032355~response to estradiol,go:0032743~positive regulation of interleukin-2 production,go:0032819~positive regulation of natural killer cell proliferation,go:0032825~positive regulation of natural killer cell differentiation,go:0032870~cellular response to hormone stimulus,go:0033028~myeloid cell apoptotic process,go:0033033~negative regulation of myeloid cell apoptotic process,go:0033077~t cell differentiation in thymus,go:0040014~regulation of multicellular organism growth,go:0040018~positive regulation of multicellular organism growth,go:0042104~positive regulation of activated t cell proliferation,go:0042127~regulation of cell proliferation,go:0042267~natural killer cell mediated cytotoxicity,go:0042448~progesterone metabolic process,go:0042492~gamma-delta t cell differentiation,go:0043029~t cell homeostasis,go:0043066~negative regulation of apoptotic process,go:0043434~response to peptide hormone,go:0045579~positive regulation of b cell differentiation,go:0045588~positive regulation of gamma-delta t cell differentiation,go:0045647~negative regulation of erythrocyte differentiation,go:0045648~positive regulation of erythrocyte differentiation,go:0045931~positive regulation of mitotic cell cycle,go:0045944~positive regulation of transcription from rna polymerase ii promoter,go:0045954~positive regulation of natural killer cell mediated cytotoxicity,go:0046543~development of secondary female sexual characteristics,go:0046544~development of secondary male sexual characteristics,go:0048541~peyer's patch development,go:0050729~positive regulation of inflammatory response,go:0050798~activated t cell proliferation,go:0060397~jak-stat cascade involved in growth hormone signaling pathway,go:0070669~response to interleukin-2,go:0070670~response to interleukin-4,go:0070672~response to interleukin-15,go:0071363~cellular response to growth factor stimulus,go:0071364~cellular response to epidermal growth factor stimulus,go:0097531~mast cell migration,\"\n",
      " 'steroid metabolic process,go:0008643~carbohydrate transport,go:0009749~response to glucose,go:0014070~response to organic cyclic compound,go:0015760~glucose-6-phosphate transport,go:0030593~neutrophil chemotaxis,go:0032682~negative regulation of chemokine production,go:0035166~post-embryonic hemopoiesis,go:0035435~phosphate ion transmembrane transport,go:0042593~glucose homeostasis,go:0042632~cholesterol homeostasis,go:0043085~positive regulation of catalytic activity,go:0045730~respiratory burst,'\n",
      " 'steroid biosynthetic process,go:0006699~bile acid biosynthetic process,go:0006701~progesterone biosynthetic process,go:0006869~lipid transport,go:0007031~peroxisome organization,go:0008206~bile acid metabolic process,go:0015914~phospholipid transport,go:0032367~intracellular cholesterol transport,go:0032385~positive regulation of intracellular cholesterol transport,go:0032959~inositol trisphosphate biosynthetic process,go:0033540~fatty acid beta-oxidation using acyl-coa oxidase,go:0036109~alpha-linolenic acid metabolic process,go:0045940~positive regulation of steroid metabolic process,go:0071071~regulation of phospholipid biosynthetic process,go:0072659~protein localization to plasma membrane,go:1901373~lipid hydroperoxide transport,positive regulation of steroid metabolic process,go:0071071~regulation of phospholipid biosynthetic process,go:0072659~protein localization to plasma membrane,go:1901373~lipid hydroperoxide transport,'\n",
      " 'response to steroid hormone,go:0048565~digestive tract development,go:0048661~positive regulation of smooth muscle cell proliferation,go:0048701~embryonic cranial skeleton morphogenesis,go:0051138~positive regulation of nk t cell differentiation,go:0060044~negative regulation of cardiac muscle cell proliferation,go:0060389~pathway-restricted smad protein phosphorylation,go:0060412~ventricular septum morphogenesis,go:0060434~bronchus morphogenesis,go:0060440~trachea formation,go:0060443~mammary gland morphogenesis,go:0060463~lung lobe morphogenesis,go:0061520~langerhans cell differentiation,go:0070723~response to cholesterol,go:0071363~cellular response to growth factor stimulus,go:0072091~regulation of stem cell proliferation,go:1905007~positive regulation of epithelial to mesenchymal transition involved in endocardial cushion formation,go:1905317~inferior endocardial cushion morphogenesis,go:1990086~lens fiber cell apoptotic process,go:1990428~mirna transport,go:2000379~positive regulation of reactive oxygen species metabolic process,go:2000563~positive regulation of cd4-positive, alpha-beta t cell proliferation,']\n"
     ]
    }
   ],
   "source": [
    "# for c in comps:\n",
    "for c in ['cytokinesis']:\n",
    "    print(gene_cats_bpcc['UP_KW_BIOLOGICAL_PROCESS'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '').unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d80e3f38",
   "metadata": {},
   "source": [
    "#### For LUAD dataset:\n",
    "\n",
    "1 - Read predictability map of categorical features (using MLP model)\n",
    "\n",
    "2 - Assign the feature categories to compartments/stains\n",
    "\n",
    "3 - Read functional annotations of the reference set according to DAVIDs output and add columns for each channel\n",
    "   - Add channel specific annotation to each columns channel\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "00bc67ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA\n",
      "0\n",
      "37\n",
      "5.224390243902439\n",
      "RNA\n",
      "0\n",
      "52\n",
      "1.3362175525339925\n",
      "AGP\n",
      "0\n",
      "32\n",
      "1.3675937122128174\n",
      "Mito\n",
      "0\n",
      "46\n",
      "1.1570552147239264\n",
      "ER\n",
      "1\n",
      "48\n",
      "1.5714285714285714\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds ratio</th>\n",
       "      <th>med_restComp_oddsratio</th>\n",
       "      <th>restComp_oddsratio</th>\n",
       "      <th>anyComp_oddsratio</th>\n",
       "      <th>top-ratio</th>\n",
       "      <th>dif</th>\n",
       "      <th>dif2</th>\n",
       "      <th>dif3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>0.719728</td>\n",
       "      <td>1.289698</td>\n",
       "      <td>6.952986</td>\n",
       "      <td>5.22439</td>\n",
       "      <td>4.06</td>\n",
       "      <td>-6.233258</td>\n",
       "      <td>-0.56997</td>\n",
       "      <td>-4.504663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA</th>\n",
       "      <td>1.375026</td>\n",
       "      <td>0.95696</td>\n",
       "      <td>1.164773</td>\n",
       "      <td>1.336218</td>\n",
       "      <td>5.7</td>\n",
       "      <td>0.210253</td>\n",
       "      <td>0.418066</td>\n",
       "      <td>0.038808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGP</th>\n",
       "      <td>0.74052</td>\n",
       "      <td>0.765185</td>\n",
       "      <td>0.9744</td>\n",
       "      <td>1.367594</td>\n",
       "      <td>3.51</td>\n",
       "      <td>-0.23388</td>\n",
       "      <td>-0.024665</td>\n",
       "      <td>-0.627074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mito</th>\n",
       "      <td>0.895629</td>\n",
       "      <td>0.965644</td>\n",
       "      <td>0.946092</td>\n",
       "      <td>1.157055</td>\n",
       "      <td>5.04</td>\n",
       "      <td>-0.050463</td>\n",
       "      <td>-0.070015</td>\n",
       "      <td>-0.261426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER</th>\n",
       "      <td>0.598337</td>\n",
       "      <td>0.987609</td>\n",
       "      <td>1.069565</td>\n",
       "      <td>1.571429</td>\n",
       "      <td>5.26</td>\n",
       "      <td>-0.471229</td>\n",
       "      <td>-0.389272</td>\n",
       "      <td>-0.973092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     odds ratio med_restComp_oddsratio restComp_oddsratio anyComp_oddsratio  \\\n",
       "DNA    0.719728               1.289698           6.952986           5.22439   \n",
       "RNA    1.375026                0.95696           1.164773          1.336218   \n",
       "AGP     0.74052               0.765185             0.9744          1.367594   \n",
       "Mito   0.895629               0.965644           0.946092          1.157055   \n",
       "ER     0.598337               0.987609           1.069565          1.571429   \n",
       "\n",
       "     top-ratio       dif      dif2      dif3  \n",
       "DNA       4.06 -6.233258  -0.56997 -4.504663  \n",
       "RNA        5.7  0.210253  0.418066  0.038808  \n",
       "AGP       3.51  -0.23388 -0.024665 -0.627074  \n",
       "Mito      5.04 -0.050463 -0.070015 -0.261426  \n",
       "ER        5.26 -0.471229 -0.389272 -0.973092  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../utils/') \n",
    "from readProfiles import rename_affyprobe_to_genename\n",
    "from saveAsNewSheetToExistingFile import saveAsNewSheetToExistingFile\n",
    "########### 1 ###########\n",
    "filename='../results/SingleGenePred_cpCategoryMap/cat_scores_maps.xlsx'\n",
    "saved_scores=pd.read_excel(filename, sheet_name=None)\n",
    "# which_ds_model='LUAD-9-MLP-ht'\n",
    "which_ds_model='LUAD-9-MLP-keras-ht'\n",
    "# which_ds_model='LUAD-9-lasso-ht'\n",
    "dfcats=saved_scores[which_ds_model].rename(columns={'Unnamed: 0':'ID'})\n",
    "dfcats=dfcats[dfcats.columns[~dfcats.isna().any()].tolist()]\n",
    "dfcats2,_=rename_affyprobe_to_genename(dfcats.set_index('ID').T,dfcats.ID.tolist())\n",
    "dfcats=dfcats2.T.reset_index()\n",
    "\n",
    "# saved_scores.keys()\n",
    "\n",
    "########### 2 ###########\n",
    "Channelss=['DNA','RNA','AGP','Mito','ER']\n",
    "Channelss_cats=['DNA|Nuclei_AreaShape','RNA','AGP|Cytoplasm_AreaShape|Cells_AreaShape','Mito','ER']\n",
    "# Channelss_cats=['DNA|Nuclei_AreaShape','RNA','AGP|Cytoplasm_AreaShape','Mito','ER']\n",
    "\n",
    "for ci in range(len(Channelss)):\n",
    "    dfcats['max_'+Channelss[ci]]=dfcats.loc[:,dfcats.columns.str.contains(Channelss_cats[ci])].max(axis=1)\n",
    "\n",
    "    \n",
    "dfcats['top_channel']=dfcats[['max_'+Channelss[ci] for ci in range(len(Channelss))]].idxmax(axis=\"columns\")\n",
    "########### 3 ###########\n",
    "# gene_cats_bpcc=pd.read_csv('./go_BP_CC_MF_DIRECT_921.txt',delimiter='\\t')\n",
    "gene_cats_bpcc=pd.read_csv('./source/GO_bp_cc_mf_direct_LUAD_976.txt',delimiter='\\t')\n",
    "comps=['mitochondri','golgi','membrane','cytoskeleton','actin','endoplasmic','rna','nucleol',\\\n",
    "       'cell division','mitosis','mitotic','cell cycle']#,'cytokine','hormone']\n",
    "#    UP_KW_BIOLOGICAL_PROCESS bad\n",
    "# UP_KW_CELLULAR_COMPONENT\n",
    "# UP_KW_MOLECULAR_FUNCTION #null\n",
    "\n",
    "# GOTERM_BP_DIRECT\n",
    "# GOTERM_CC_DIRECT\n",
    "# GOTERM_MF_DIRECT\n",
    "for c in comps:\n",
    "    gene_cats_bpcc[c]=gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_MF_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')#+\\\n",
    "#     gene_cats_bpcc['UP_KW_BIOLOGICAL_PROCESS'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "#     gene_cats_bpcc['UP_KW_CELLULAR_COMPONENT'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "#     gene_cats_bpcc['UP_KW_MOLECULAR_FUNCTION'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')#+\\\n",
    "# #     gene_cats_bpcc['UP_SEQ_FEATURE'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split(':')[-1]+','+x[si:].split(',')[0] for si in locations_of_substring(x,c)]) if c in x else '')\n",
    "    \n",
    "gene_cats_bpcc['RNA_nucleoli']=gene_cats_bpcc['rna']+gene_cats_bpcc['nucleol']\n",
    "gene_cats_bpcc['DNA']=gene_cats_bpcc['cell division']+gene_cats_bpcc['mitosis']+\\\n",
    "gene_cats_bpcc['mitotic']+gene_cats_bpcc['cell cycle']\n",
    "\n",
    "gene_cats_bpcc['cytoskeleton-actin']=gene_cats_bpcc['cytoskeleton']+gene_cats_bpcc['actin']+\\\n",
    "gene_cats_bpcc['golgi']+gene_cats_bpcc['membrane']#+gene_cats_bpcc['cytokine']\n",
    "# gene_cats_bpcc['ER']=gene_cats_bpcc['endoplasmic']\n",
    "gene_cats_bpcc['mitochondria']=gene_cats_bpcc['mitochondri']#+gene_cats_bpcc['hormone']\n",
    "\n",
    "gene_cats_bpcc=pd.merge(gene_cats_bpcc,dfcats,how='inner',on='ID')\n",
    "\n",
    "#########################\n",
    "Channelss_dict={'DNA':'DNA','RNA_nucleoli':'RNA','cytoskeleton-actin':'AGP','mitochondria':'Mito','endoplasmic':'ER'}\n",
    "Chan_rev_dict = dict(zip(Channelss_dict.values(),Channelss_dict.keys()))\n",
    "\n",
    "comps2=['mitochondri','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "gene_cats_bpcc['any_comps']=''\n",
    "for co in comps2:\n",
    "    gene_cats_bpcc['any_comps']=gene_cats_bpcc['any_comps']+gene_cats_bpcc[co]\n",
    "        \n",
    "        \n",
    "        \n",
    "# gene_cats_bpcc=gene_cats_bpcc[gene_cats_bpcc['any_comps']!=''].reset_index(drop=True)        \n",
    "from sklearn.metrics import confusion_matrix\n",
    "from scipy.stats import fisher_exact\n",
    "# top_bool=(gene_cats_bpcc['top58']==True).values\n",
    "table2=pd.DataFrame(index=Channelss_dict.keys(),columns=Channelss_dict.values())\n",
    "table3=pd.DataFrame(index=Channelss,columns=['odds ratio','med_restComp_oddsratio','restComp_oddsratio','anyComp_oddsratio','top-ratio'])\n",
    "# table3=pd.DataFrame(index=Channelss,columns=['Prevalence','anyComp-Prevalence','noComp-Prevalence','top-ratio'])\n",
    "\n",
    "table=[]\n",
    "for c in Channelss:\n",
    "    print(c)\n",
    "    c_r=list(set(Channelss)-set([c]))\n",
    "#     gene_cats_bpcc_highP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.7)]\n",
    "#     gene_cats_bpcc_lowP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]<0.1)]\n",
    "    \n",
    "    low_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[:100].index\n",
    "    high_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[-100:].index    \n",
    "    gene_cats_bpcc_highP=gene_cats_bpcc.loc[high_ind].reset_index(drop=True)\n",
    "    gene_cats_bpcc_lowP=gene_cats_bpcc.loc[low_ind].reset_index(drop=True)\n",
    "    \n",
    "#     gene_cats_bpcc_highP[]\n",
    "    \n",
    "#     gene_cats_bpcc2=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.5) | (gene_cats_bpcc['max_'+c]<0)].reset_index(drop=True)\n",
    "    gene_cats_bpcc2=gene_cats_bpcc.copy()\n",
    "#     top_bool=(gene_cats_bpcc2['max_'+c]>0.5).values\n",
    "    \n",
    "#     gene_cats_bpcc_highP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.5)]\n",
    "#     gene_cats_bpcc_lowP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]<0)]    \n",
    "    \n",
    "#     top_bool=(gene_cats_bpcc['max_'+c]>.6).values\n",
    "    top_bool=(gene_cats_bpcc['max_'+c]<0).values\n",
    "    \n",
    "    print(np.sum(gene_cats_bpcc.loc[gene_cats_bpcc['max_'+c]<0,['max_'+Channelss[ci] for ci in range(len(Channelss))]].max(axis=1)>0.3))\n",
    "#     top_bool=(gene_cats_bpcc['max_'+c]>0.6).values\n",
    "#     top_bool=((gene_cats_bpcc['top_channel']=='max_'+c) &\\\n",
    "#               (gene_cats_bpcc['max_'+c]>0.4) &\\\n",
    "#              (gene_cats_bpcc[['max_'+cr for cr in c_r]].max(axis=1)<0.4)).values\n",
    "#     top_bool=((gene_cats_bpcc['max_'+c]-gene_cats_bpcc[['max_'+cr for cr in c_r]].max(axis=1)>0.05) &\\\n",
    "#              (gene_cats_bpcc['max_'+c]>0.2)).values    \n",
    "    \n",
    "#     index_of_max=gene_cats_bpcc[(gene_cats_bpcc['top_channel']=='max_'+c)].sort_values(by='max_'+c,ascending=True)\\\n",
    "#     .index[0:45].values\n",
    "#     top_bool=np.zeros(gene_cats_bpcc.shape[0],).astype(bool)\n",
    "#     top_bool[index_of_max]=True    \n",
    "    \n",
    "#     hhgj\n",
    "    n_top=sum(top_bool)\n",
    "    print(n_top)\n",
    "    \n",
    "    comps2=['mitochondria','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "    table1=pd.DataFrame(index=comps2+['any comp','no comp'],columns=['Prevalence','p-value','odds ratio'])\n",
    "    for co in comps2:\n",
    "#         ratio_h=gene_cats_bpcc_highP[(gene_cats_bpcc_highP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_highP.shape[0]\n",
    "\n",
    "#         ratio_l=gene_cats_bpcc_lowP[(gene_cats_bpcc_lowP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_lowP.shape[0]\n",
    "# #         fdsfsd\n",
    "#         print(c,'-',co,': ',ratio_h/ratio_l,gene_cats_bpcc_highP.shape[0],gene_cats_bpcc_lowP.shape[0])\n",
    "        \n",
    "        enr_ratio=gene_cats_bpcc2[top_bool & (gene_cats_bpcc2[co]!='')].shape[0]/\\\n",
    "        gene_cats_bpcc2[(gene_cats_bpcc2[co]!='')].shape[0]\n",
    "        comp_bool=(gene_cats_bpcc2[co]!='').values\n",
    "        oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, comp_bool))\n",
    "#         gkkh\n",
    "    #     print(co, ':',gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc[co]!='')].shape[0],\\\n",
    "    #          ', ',gene_cats_bpcc[(gene_cats_bpcc[co]!='')].shape[0])\n",
    "        \n",
    "        table1.loc[co,['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "#         print(co,\": \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "        table2.loc[co,c]=oddsratio\n",
    "         \n",
    "    \n",
    "    enr_ratio=gene_cats_bpcc[top_bool & (gene_cats_bpcc['any_comps']!='')].shape[0]/\\\n",
    "    gene_cats_bpcc[(gene_cats_bpcc['any_comps']!='')].shape[0]\n",
    "    any_oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['any_comps']!='').values))\n",
    "#     print(\"any_comps: \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "    table1.loc['any comp',['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,any_oddsratio\n",
    "    print(any_oddsratio)\n",
    "\n",
    "    nocomp_enr_ratio=gene_cats_bpcc[(top_bool) & (gene_cats_bpcc['any_comps']=='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['any_comps']=='')].shape[0]\n",
    "\n",
    "    comps2=['mitochondria','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "    comps2.remove(Chan_rev_dict[c])\n",
    "    \n",
    "    med_restComp_oddsratio=table2.loc[comps2,c].median()\n",
    "    gene_cats_bpcc['rest_comps']=''\n",
    "    for co in comps2:\n",
    "        gene_cats_bpcc['rest_comps']=gene_cats_bpcc['rest_comps']+gene_cats_bpcc[co]    \n",
    "    \n",
    "    rest_enr_ratio=gene_cats_bpcc[(top_bool) & (gene_cats_bpcc['rest_comps']=='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['rest_comps']=='')].shape[0]    \n",
    "    rest_oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['rest_comps']!='').values))\n",
    "    \n",
    "    table.append(table1)\n",
    "#     print('num top ('+ str(n_top)+')/total genes (912): ', np.round((n_top/912)*100,2),'%')\n",
    "    table3.loc[c,['odds ratio','med_restComp_oddsratio','restComp_oddsratio','anyComp_oddsratio','top-ratio']]=\\\n",
    "    table1.loc[Chan_rev_dict[c],'odds ratio'],med_restComp_oddsratio,rest_oddsratio,\\\n",
    "    any_oddsratio,np.round((n_top/912)*100,2)\n",
    "    \n",
    "#     table3.loc[c,['Prevalence','anyComp-Prevalence','noComp-Prevalence','top-ratio']]=table1.loc[Chan_rev_dict[c],'Prevalence'],\\\n",
    "#     enr_ratio*100,nocomp_enr_ratio*100,np.round((n_top/912)*100,2)    \n",
    "    \n",
    "#     print(Chan_rev_dict[c],': ',table1.loc[Chan_rev_dict[c],['Prevalence']].values)\n",
    "    \n",
    "# # table3['dif']=table3['Prevalence']-table3['anyComp-Prevalence']\n",
    "table3['dif']=table3['odds ratio']-table3['restComp_oddsratio']\n",
    "table3['dif2']=table3['odds ratio']-table3['med_restComp_oddsratio']\n",
    "table3['dif3']=table3['odds ratio']-table3['anyComp_oddsratio']\n",
    "# print(table3['dif'].min(),table3['dif'].sum())\n",
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d5d01fce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a1=(gene_cats_bpcc2[top_bool & (gene_cats_bpcc2[co]!='')].shape[0])/(gene_cats_bpcc2[(top_bool) & (gene_cats_bpcc2[co]=='')].shape[0])\n",
    "# a2=(gene_cats_bpcc2[(~top_bool) & (gene_cats_bpcc2[co]!='')].shape[0])/(gene_cats_bpcc2[(~top_bool) & (gene_cats_bpcc2[co]=='')].shape[0])\n",
    "\n",
    "# table3\n",
    "# gene_cats_bpcc[(gene_cats_bpcc['top_channel']=='max_'+c)].sort_values(by='max_'+c,ascending=False)\n",
    "# table1\n",
    "# \n",
    "# gene_cats_bpcc[gene_cats_bpcc['max_'+c]<0]\n",
    "np.sum(gene_cats_bpcc.loc[gene_cats_bpcc['max_'+c]<0,['max_'+Channelss[ci] for ci in range(len(Channelss))]].max(axis=1)>0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "8169351d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# high_ind.shape\n",
    "# gene_cats_bpcc_lowP[(gene_cats_bpcc_lowP[co]!='')]\n",
    "# table3\n",
    "# a1/a2\n",
    "# gene_cats_bpcc[gene_cats_bpcc['any_comps']!=''].reset_index(drop=True)#.shape\n",
    "x1=table3.astype(float).round(3)[['odds ratio','restComp_oddsratio']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d3d314d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1e-10"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1e-10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "4275ebd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ExtendedData5  saved!\n"
     ]
    }
   ],
   "source": [
    "# table2.loc[comps2].median()\n",
    "# pd.concat([table3.astype(float).round(3)[['odds ratio','restComp_oddsratio']],x1],axis=1)\n",
    "\n",
    "source_data_add='../Figs_Source_Data.xlsx'\n",
    "if 0:\n",
    "    saveAsNewSheetToExistingFile(source_data_add,pd.concat([table3.astype(float).round(3)[['odds ratio','restComp_oddsratio']],x1],axis=1),'ExtendedData5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "id": "8194f59a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>GOTERM_BP_DIRECT</th>\n",
       "      <th>GOTERM_CC_DIRECT</th>\n",
       "      <th>GOTERM_MF_DIRECT</th>\n",
       "      <th>KEGG_PATHWAY</th>\n",
       "      <th>UP_KW_BIOLOGICAL_PROCESS</th>\n",
       "      <th>UP_KW_CELLULAR_COMPONENT</th>\n",
       "      <th>UP_KW_MOLECULAR_FUNCTION</th>\n",
       "      <th>...</th>\n",
       "      <th>Cytoplasm_AreaShape</th>\n",
       "      <th>Cells_AreaShape</th>\n",
       "      <th>max_DNA</th>\n",
       "      <th>max_RNA</th>\n",
       "      <th>max_AGP</th>\n",
       "      <th>max_Mito</th>\n",
       "      <th>max_ER</th>\n",
       "      <th>top_channel</th>\n",
       "      <th>any_comps</th>\n",
       "      <th>rest_comps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DECR1</td>\n",
       "      <td>2,4-dienoyl-CoA reductase 1(DECR1)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0006635~fatty acid beta-oxidation,</td>\n",
       "      <td>GO:0005634~nucleus,GO:0005654~nucleoplasm,GO:0...</td>\n",
       "      <td>GO:0008670~2,4-dienoyl-CoA reductase (NADPH) a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0276~Fatty acid metabolism,KW-0443~Lipid me...</td>\n",
       "      <td>KW-0496~Mitochondrion,</td>\n",
       "      <td>KW-0560~Oxidoreductase,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.179192</td>\n",
       "      <td>0.242334</td>\n",
       "      <td>0.259064</td>\n",
       "      <td>0.272621</td>\n",
       "      <td>0.333014</td>\n",
       "      <td>0.302037</td>\n",
       "      <td>0.398891</td>\n",
       "      <td>max_ER</td>\n",
       "      <td>mitochondrion,mitochondrial matrix,</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADO</td>\n",
       "      <td>2-aminoethanethiol dioxygenase(ADO)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0071456~cellular response to hypoxia,</td>\n",
       "      <td>GO:0005829~cytosol,</td>\n",
       "      <td>GO:0005506~iron ion binding,GO:0005515~protein...</td>\n",
       "      <td>hsa00430:Taurine and hypotaurine metabolism,hs...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0223~Dioxygenase,KW-0560~Oxidoreductase,</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.100508</td>\n",
       "      <td>-0.042940</td>\n",
       "      <td>-0.030797</td>\n",
       "      <td>-0.018856</td>\n",
       "      <td>-0.042940</td>\n",
       "      <td>0.010047</td>\n",
       "      <td>-0.004270</td>\n",
       "      <td>max_Mito</td>\n",
       "      <td>oxidoreductase activity, acting on single dono...</td>\n",
       "      <td>oxidoreductase activity, acting on single dono...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HMGCR</td>\n",
       "      <td>3-hydroxy-3-methylglutaryl-CoA reductase(HMGCR)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0006695~cholesterol biosynthetic process,GO...</td>\n",
       "      <td>GO:0005778~peroxisomal membrane,GO:0005783~end...</td>\n",
       "      <td>GO:0004420~hydroxymethylglutaryl-CoA reductase...</td>\n",
       "      <td>hsa00900:Terpenoid backbone biosynthesis,hsa01...</td>\n",
       "      <td>KW-0152~Cholesterol biosynthesis,KW-0153~Chole...</td>\n",
       "      <td>KW-0256~Endoplasmic reticulum,KW-0472~Membrane...</td>\n",
       "      <td>KW-0560~Oxidoreductase,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.175751</td>\n",
       "      <td>0.176601</td>\n",
       "      <td>0.312376</td>\n",
       "      <td>0.308959</td>\n",
       "      <td>0.258915</td>\n",
       "      <td>0.342495</td>\n",
       "      <td>0.254575</td>\n",
       "      <td>max_Mito</td>\n",
       "      <td>peroxisomal membrane,endoplasmic reticulum mem...</td>\n",
       "      <td>peroxisomal membrane,endoplasmic reticulum mem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HMGCS1</td>\n",
       "      <td>3-hydroxy-3-methylglutaryl-CoA synthase 1(HMGCS1)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0001889~liver development,GO:0006084~acetyl...</td>\n",
       "      <td>GO:0005737~cytoplasm,GO:0005829~cytosol,</td>\n",
       "      <td>GO:0004421~hydroxymethylglutaryl-CoA synthase ...</td>\n",
       "      <td>hsa00280:Valine, leucine and isoleucine degrad...</td>\n",
       "      <td>KW-0152~Cholesterol biosynthesis,KW-0153~Chole...</td>\n",
       "      <td>KW-0963~Cytoplasm,</td>\n",
       "      <td>KW-0808~Transferase,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.495392</td>\n",
       "      <td>0.481254</td>\n",
       "      <td>0.566847</td>\n",
       "      <td>0.502503</td>\n",
       "      <td>0.515804</td>\n",
       "      <td>0.567605</td>\n",
       "      <td>0.569166</td>\n",
       "      <td>max_ER</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HACD3</td>\n",
       "      <td>3-hydroxyacyl-CoA dehydratase 3(HACD3)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0006633~fatty acid biosynthetic process,GO:...</td>\n",
       "      <td>GO:0005783~endoplasmic reticulum,GO:0005789~en...</td>\n",
       "      <td>GO:0005096~GTPase activator activity,GO:000551...</td>\n",
       "      <td>hsa00062:Fatty acid elongation,hsa01040:Biosyn...</td>\n",
       "      <td>KW-0275~Fatty acid biosynthesis,KW-0276~Fatty ...</td>\n",
       "      <td>KW-0256~Endoplasmic reticulum,KW-0472~Membrane,</td>\n",
       "      <td>KW-0456~Lyase,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040313</td>\n",
       "      <td>-0.051127</td>\n",
       "      <td>0.063998</td>\n",
       "      <td>0.070532</td>\n",
       "      <td>0.040313</td>\n",
       "      <td>0.086491</td>\n",
       "      <td>0.179419</td>\n",
       "      <td>max_ER</td>\n",
       "      <td>endoplasmic reticulum membrane,integral compon...</td>\n",
       "      <td>endoplasmic reticulum membrane,integral compon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>ZNF395</td>\n",
       "      <td>zinc finger protein 395(ZNF395)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0006355~regulation of transcription, DNA-te...</td>\n",
       "      <td>GO:0005634~nucleus,GO:0005654~nucleoplasm,GO:0...</td>\n",
       "      <td>GO:0000978~RNA polymerase II core promoter pro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0804~Transcription,KW-0805~Transcription re...</td>\n",
       "      <td>KW-0539~Nucleus,KW-0963~Cytoplasm,</td>\n",
       "      <td>KW-0238~DNA-binding,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.104552</td>\n",
       "      <td>0.152037</td>\n",
       "      <td>0.189749</td>\n",
       "      <td>0.151408</td>\n",
       "      <td>0.152037</td>\n",
       "      <td>0.112444</td>\n",
       "      <td>0.140626</td>\n",
       "      <td>max_DNA</td>\n",
       "      <td>rna polymerase ii core promoter proximal regio...</td>\n",
       "      <td>rna polymerase ii core promoter proximal regio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>ZNF451</td>\n",
       "      <td>zinc finger protein 451(ZNF451)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0010468~regulation of gene expression,GO:00...</td>\n",
       "      <td>GO:0005634~nucleus,GO:0016021~integral compone...</td>\n",
       "      <td>GO:0003714~transcription corepressor activity,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0804~Transcription,KW-0805~Transcription re...</td>\n",
       "      <td>KW-0472~Membrane,KW-0539~Nucleus,</td>\n",
       "      <td>KW-0808~Transferase,</td>\n",
       "      <td>...</td>\n",
       "      <td>0.411780</td>\n",
       "      <td>0.336247</td>\n",
       "      <td>0.522736</td>\n",
       "      <td>0.523008</td>\n",
       "      <td>0.485401</td>\n",
       "      <td>0.567924</td>\n",
       "      <td>0.462915</td>\n",
       "      <td>max_Mito</td>\n",
       "      <td>integral component of membrane,positive regula...</td>\n",
       "      <td>integral component of membrane,positive regula...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>ZNF586</td>\n",
       "      <td>zinc finger protein 586(ZNF586)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0006355~regulation of transcription, DNA-te...</td>\n",
       "      <td>GO:0005634~nucleus,</td>\n",
       "      <td>GO:0000978~RNA polymerase II core promoter pro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0804~Transcription,KW-0805~Transcription re...</td>\n",
       "      <td>KW-0539~Nucleus,</td>\n",
       "      <td>KW-0238~DNA-binding,</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.002468</td>\n",
       "      <td>-0.033174</td>\n",
       "      <td>0.014831</td>\n",
       "      <td>-0.064294</td>\n",
       "      <td>-0.002468</td>\n",
       "      <td>-0.031964</td>\n",
       "      <td>0.008457</td>\n",
       "      <td>max_DNA</td>\n",
       "      <td>rna polymerase ii core promoter proximal regio...</td>\n",
       "      <td>rna polymerase ii core promoter proximal regio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>ZNF589</td>\n",
       "      <td>zinc finger protein 589(ZNF589)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0000122~negative regulation of transcriptio...</td>\n",
       "      <td>GO:0005634~nucleus,GO:0005654~nucleoplasm,</td>\n",
       "      <td>GO:0000977~RNA polymerase II regulatory region...</td>\n",
       "      <td>hsa05168:Herpes simplex virus 1 infection,</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0539~Nucleus,</td>\n",
       "      <td>KW-0238~DNA-binding,KW-0678~Repressor,</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.028800</td>\n",
       "      <td>-0.056366</td>\n",
       "      <td>0.035850</td>\n",
       "      <td>-0.008191</td>\n",
       "      <td>0.006670</td>\n",
       "      <td>0.040527</td>\n",
       "      <td>-0.026931</td>\n",
       "      <td>max_Mito</td>\n",
       "      <td>rna polymerase ii regulatory region sequence-s...</td>\n",
       "      <td>rna polymerase ii regulatory region sequence-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>ZW10</td>\n",
       "      <td>zw10 kinetochore protein(ZW10)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GO:0000070~mitotic sister chromatid segregatio...</td>\n",
       "      <td>GO:0000776~kinetochore,GO:0000922~spindle pole...</td>\n",
       "      <td>GO:0005515~protein binding,GO:0019237~centrome...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0131~Cell cycle,KW-0132~Cell division,KW-04...</td>\n",
       "      <td>KW-0137~Centromere,KW-0158~Chromosome,KW-0206~...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.070375</td>\n",
       "      <td>0.044094</td>\n",
       "      <td>0.077144</td>\n",
       "      <td>0.025763</td>\n",
       "      <td>0.070375</td>\n",
       "      <td>0.051858</td>\n",
       "      <td>0.124461</td>\n",
       "      <td>max_ER</td>\n",
       "      <td>er to retrograde vesicle-mediated transport, e...</td>\n",
       "      <td>er to retrograde vesicle-mediated transport, e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>976 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                                          Gene Name       Species  \\\n",
       "0     DECR1                 2,4-dienoyl-CoA reductase 1(DECR1)  Homo sapiens   \n",
       "1       ADO                2-aminoethanethiol dioxygenase(ADO)  Homo sapiens   \n",
       "2     HMGCR    3-hydroxy-3-methylglutaryl-CoA reductase(HMGCR)  Homo sapiens   \n",
       "3    HMGCS1  3-hydroxy-3-methylglutaryl-CoA synthase 1(HMGCS1)  Homo sapiens   \n",
       "4     HACD3             3-hydroxyacyl-CoA dehydratase 3(HACD3)  Homo sapiens   \n",
       "..      ...                                                ...           ...   \n",
       "971  ZNF395                    zinc finger protein 395(ZNF395)  Homo sapiens   \n",
       "972  ZNF451                    zinc finger protein 451(ZNF451)  Homo sapiens   \n",
       "973  ZNF586                    zinc finger protein 586(ZNF586)  Homo sapiens   \n",
       "974  ZNF589                    zinc finger protein 589(ZNF589)  Homo sapiens   \n",
       "975    ZW10                     zw10 kinetochore protein(ZW10)  Homo sapiens   \n",
       "\n",
       "                                      GOTERM_BP_DIRECT  \\\n",
       "0                GO:0006635~fatty acid beta-oxidation,   \n",
       "1             GO:0071456~cellular response to hypoxia,   \n",
       "2    GO:0006695~cholesterol biosynthetic process,GO...   \n",
       "3    GO:0001889~liver development,GO:0006084~acetyl...   \n",
       "4    GO:0006633~fatty acid biosynthetic process,GO:...   \n",
       "..                                                 ...   \n",
       "971  GO:0006355~regulation of transcription, DNA-te...   \n",
       "972  GO:0010468~regulation of gene expression,GO:00...   \n",
       "973  GO:0006355~regulation of transcription, DNA-te...   \n",
       "974  GO:0000122~negative regulation of transcriptio...   \n",
       "975  GO:0000070~mitotic sister chromatid segregatio...   \n",
       "\n",
       "                                      GOTERM_CC_DIRECT  \\\n",
       "0    GO:0005634~nucleus,GO:0005654~nucleoplasm,GO:0...   \n",
       "1                                  GO:0005829~cytosol,   \n",
       "2    GO:0005778~peroxisomal membrane,GO:0005783~end...   \n",
       "3             GO:0005737~cytoplasm,GO:0005829~cytosol,   \n",
       "4    GO:0005783~endoplasmic reticulum,GO:0005789~en...   \n",
       "..                                                 ...   \n",
       "971  GO:0005634~nucleus,GO:0005654~nucleoplasm,GO:0...   \n",
       "972  GO:0005634~nucleus,GO:0016021~integral compone...   \n",
       "973                                GO:0005634~nucleus,   \n",
       "974         GO:0005634~nucleus,GO:0005654~nucleoplasm,   \n",
       "975  GO:0000776~kinetochore,GO:0000922~spindle pole...   \n",
       "\n",
       "                                      GOTERM_MF_DIRECT  \\\n",
       "0    GO:0008670~2,4-dienoyl-CoA reductase (NADPH) a...   \n",
       "1    GO:0005506~iron ion binding,GO:0005515~protein...   \n",
       "2    GO:0004420~hydroxymethylglutaryl-CoA reductase...   \n",
       "3    GO:0004421~hydroxymethylglutaryl-CoA synthase ...   \n",
       "4    GO:0005096~GTPase activator activity,GO:000551...   \n",
       "..                                                 ...   \n",
       "971  GO:0000978~RNA polymerase II core promoter pro...   \n",
       "972  GO:0003714~transcription corepressor activity,...   \n",
       "973  GO:0000978~RNA polymerase II core promoter pro...   \n",
       "974  GO:0000977~RNA polymerase II regulatory region...   \n",
       "975  GO:0005515~protein binding,GO:0019237~centrome...   \n",
       "\n",
       "                                          KEGG_PATHWAY  \\\n",
       "0                                                  NaN   \n",
       "1    hsa00430:Taurine and hypotaurine metabolism,hs...   \n",
       "2    hsa00900:Terpenoid backbone biosynthesis,hsa01...   \n",
       "3    hsa00280:Valine, leucine and isoleucine degrad...   \n",
       "4    hsa00062:Fatty acid elongation,hsa01040:Biosyn...   \n",
       "..                                                 ...   \n",
       "971                                                NaN   \n",
       "972                                                NaN   \n",
       "973                                                NaN   \n",
       "974         hsa05168:Herpes simplex virus 1 infection,   \n",
       "975                                                NaN   \n",
       "\n",
       "                              UP_KW_BIOLOGICAL_PROCESS  \\\n",
       "0    KW-0276~Fatty acid metabolism,KW-0443~Lipid me...   \n",
       "1                                                  NaN   \n",
       "2    KW-0152~Cholesterol biosynthesis,KW-0153~Chole...   \n",
       "3    KW-0152~Cholesterol biosynthesis,KW-0153~Chole...   \n",
       "4    KW-0275~Fatty acid biosynthesis,KW-0276~Fatty ...   \n",
       "..                                                 ...   \n",
       "971  KW-0804~Transcription,KW-0805~Transcription re...   \n",
       "972  KW-0804~Transcription,KW-0805~Transcription re...   \n",
       "973  KW-0804~Transcription,KW-0805~Transcription re...   \n",
       "974                                                NaN   \n",
       "975  KW-0131~Cell cycle,KW-0132~Cell division,KW-04...   \n",
       "\n",
       "                              UP_KW_CELLULAR_COMPONENT  \\\n",
       "0                               KW-0496~Mitochondrion,   \n",
       "1                                                  NaN   \n",
       "2    KW-0256~Endoplasmic reticulum,KW-0472~Membrane...   \n",
       "3                                   KW-0963~Cytoplasm,   \n",
       "4      KW-0256~Endoplasmic reticulum,KW-0472~Membrane,   \n",
       "..                                                 ...   \n",
       "971                 KW-0539~Nucleus,KW-0963~Cytoplasm,   \n",
       "972                  KW-0472~Membrane,KW-0539~Nucleus,   \n",
       "973                                   KW-0539~Nucleus,   \n",
       "974                                   KW-0539~Nucleus,   \n",
       "975  KW-0137~Centromere,KW-0158~Chromosome,KW-0206~...   \n",
       "\n",
       "                        UP_KW_MOLECULAR_FUNCTION  ... Cytoplasm_AreaShape  \\\n",
       "0                        KW-0560~Oxidoreductase,  ...            0.179192   \n",
       "1    KW-0223~Dioxygenase,KW-0560~Oxidoreductase,  ...           -0.100508   \n",
       "2                        KW-0560~Oxidoreductase,  ...            0.175751   \n",
       "3                           KW-0808~Transferase,  ...            0.495392   \n",
       "4                                 KW-0456~Lyase,  ...            0.040313   \n",
       "..                                           ...  ...                 ...   \n",
       "971                         KW-0238~DNA-binding,  ...            0.104552   \n",
       "972                         KW-0808~Transferase,  ...            0.411780   \n",
       "973                         KW-0238~DNA-binding,  ...           -0.002468   \n",
       "974       KW-0238~DNA-binding,KW-0678~Repressor,  ...           -0.028800   \n",
       "975                                          NaN  ...            0.070375   \n",
       "\n",
       "    Cells_AreaShape   max_DNA   max_RNA   max_AGP  max_Mito    max_ER  \\\n",
       "0          0.242334  0.259064  0.272621  0.333014  0.302037  0.398891   \n",
       "1         -0.042940 -0.030797 -0.018856 -0.042940  0.010047 -0.004270   \n",
       "2          0.176601  0.312376  0.308959  0.258915  0.342495  0.254575   \n",
       "3          0.481254  0.566847  0.502503  0.515804  0.567605  0.569166   \n",
       "4         -0.051127  0.063998  0.070532  0.040313  0.086491  0.179419   \n",
       "..              ...       ...       ...       ...       ...       ...   \n",
       "971        0.152037  0.189749  0.151408  0.152037  0.112444  0.140626   \n",
       "972        0.336247  0.522736  0.523008  0.485401  0.567924  0.462915   \n",
       "973       -0.033174  0.014831 -0.064294 -0.002468 -0.031964  0.008457   \n",
       "974       -0.056366  0.035850 -0.008191  0.006670  0.040527 -0.026931   \n",
       "975        0.044094  0.077144  0.025763  0.070375  0.051858  0.124461   \n",
       "\n",
       "    top_channel                                          any_comps  \\\n",
       "0        max_ER                mitochondrion,mitochondrial matrix,   \n",
       "1      max_Mito  oxidoreductase activity, acting on single dono...   \n",
       "2      max_Mito  peroxisomal membrane,endoplasmic reticulum mem...   \n",
       "3        max_ER                                                      \n",
       "4        max_ER  endoplasmic reticulum membrane,integral compon...   \n",
       "..          ...                                                ...   \n",
       "971     max_DNA  rna polymerase ii core promoter proximal regio...   \n",
       "972    max_Mito  integral component of membrane,positive regula...   \n",
       "973     max_DNA  rna polymerase ii core promoter proximal regio...   \n",
       "974    max_Mito  rna polymerase ii regulatory region sequence-s...   \n",
       "975      max_ER  er to retrograde vesicle-mediated transport, e...   \n",
       "\n",
       "                                            rest_comps  \n",
       "0                                                       \n",
       "1    oxidoreductase activity, acting on single dono...  \n",
       "2    peroxisomal membrane,endoplasmic reticulum mem...  \n",
       "3                                                       \n",
       "4    endoplasmic reticulum membrane,integral compon...  \n",
       "..                                                 ...  \n",
       "971  rna polymerase ii core promoter proximal regio...  \n",
       "972  integral component of membrane,positive regula...  \n",
       "973  rna polymerase ii core promoter proximal regio...  \n",
       "974  rna polymerase ii regulatory region sequence-s...  \n",
       "975  er to retrograde vesicle-mediated transport, e...  \n",
       "\n",
       "[976 rows x 56 columns]"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "id": "4bacb255",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['mitochondria', 'cytoskeleton-actin', 'RNA_nucleoli', 'DNA']"
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# high_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[:-100].index\n",
    "# low_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[-100:].index\n",
    "comps2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "5061adef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mitochondria 185\n",
      "cytoskeleton-actin 617\n",
      "endoplasmic 156\n",
      "RNA_nucleoli 361\n",
      "DNA 205\n"
     ]
    }
   ],
   "source": [
    "# gene_cats_bpcc.loc[high_ind]\n",
    "comps2=['mitochondria','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "for co in comps2:\n",
    "#         ratio_h=gene_cats_bpcc_highP[(gene_cats_bpcc_highP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_highP.shape[0]\n",
    "\n",
    "#         ratio_l=gene_cats_bpcc_lowP[(gene_cats_bpcc_lowP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_lowP.shape[0]\n",
    "# #         fdsfsd\n",
    "#         print(c,'-',co,': ',ratio_h/ratio_l,gene_cats_bpcc_highP.shape[0],gene_cats_bpcc_lowP.shape[0])\n",
    "\n",
    "    print(co,gene_cats_bpcc2[(gene_cats_bpcc2[co]!='')].shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "3734d944",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([476, 241,  55, 908, 436, 149, 699, 435, 125, 299, 744, 273,  69,\n",
       "            607, 303, 605, 765, 974, 459,  92, 183, 836, 180, 352, 143, 590,\n",
       "            130, 424, 655, 775, 637, 156, 227, 832, 871,  26, 573,  90, 622,\n",
       "            575,  17,   1, 843, 550, 814, 326, 890,  45, 900, 677, 135, 569,\n",
       "            583, 909, 855, 828, 454, 280, 327, 894, 662, 973, 615, 282, 357,\n",
       "            404,  63, 120, 912, 597,  28, 861, 721, 649, 617, 751,  64, 591,\n",
       "            845, 586, 342, 155, 937, 428, 929, 713,  70, 959, 127, 527, 340,\n",
       "            341, 173, 228,  99, 383, 300, 132, 659, 792],\n",
       "           dtype='int64')"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Channelss.index(c)\n",
    "# gene_cats_bpcc2[(gene_cats_bpcc2[co]!='')].shape[0]\n",
    "gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[:100].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "210fa6d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([166, 472, 201,  50, 759, 904, 377, 687, 916, 652, 886, 621, 581,\n",
       "            728, 482, 626,  44, 896, 643, 789, 535, 588, 623, 932, 244, 154,\n",
       "            359, 369, 388, 703, 504, 666, 313, 786,  78, 641,  83, 927, 253,\n",
       "            256, 833, 372,  68, 421, 815, 532, 878, 115, 754, 438, 735,  56,\n",
       "            498, 221, 324, 730, 701, 598, 829, 555, 821, 891, 661, 401, 934,\n",
       "            289, 457, 618, 456, 370,  81, 243, 907, 955, 624, 323, 692, 638,\n",
       "            542, 443,  88, 411, 295, 817, 400, 145, 523, 260, 366, 540, 838,\n",
       "            285, 642, 939, 835, 848, 328, 948,  79, 426],\n",
       "           dtype='int64')"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[-100:].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "a12b8c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# top_bool\n",
    "# (gene_cats_bpcc['top_channel']=='max_'+c) & (gene_cats_bpcc['max_'+c]>0.6)\n",
    "# top_bool.shape\n",
    "# index_of_max\n",
    "# list(set(Channelss)-set([c]))\n",
    "# gene_cats_bpcc2=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.6) | (gene_cats_bpcc['max_'+c]<0.1)].reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "2323b6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_of_max=gene_cats_bpcc[(gene_cats_bpcc['top_channel']=='max_'+c)].sort_values(by='max_'+c,ascending=False).index[0:100].values\n",
    "top_bool=np.zeros(976,).astype(bool)\n",
    "top_bool[index_of_max]=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "8f7bd06c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# top_bool=np.zeros(976,).astype(bool)\n",
    "# top_bool[index_of_max]=True\n",
    "top_bool[427]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "68e14519",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(976, 52)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene_cats_bpcc.shape\n",
    "\n",
    "luad_genes_to_inspect_ims=['MLLT11','FAM20B','SQSTM1','TXNRD1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "8cb724bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17, 52)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc[gene_cats_bpcc['GOTERM_MF_DIRECT'].isnull()].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "31584a81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(975, 6)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('./source/GO_bp_cc_mf_direct_LUAD_975.txt',delimiter='\\t').shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f0b4ab5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds ratio</th>\n",
       "      <th>restComp_oddsratio</th>\n",
       "      <th>anyComp_oddsratio</th>\n",
       "      <th>top-ratio</th>\n",
       "      <th>dif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>2.187816</td>\n",
       "      <td>0.64314</td>\n",
       "      <td>1.025543</td>\n",
       "      <td>9.1</td>\n",
       "      <td>1.544676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA</th>\n",
       "      <td>0.952168</td>\n",
       "      <td>1.004431</td>\n",
       "      <td>0.993456</td>\n",
       "      <td>9.76</td>\n",
       "      <td>-0.052264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGP</th>\n",
       "      <td>0.851337</td>\n",
       "      <td>1.041846</td>\n",
       "      <td>0.992218</td>\n",
       "      <td>10.64</td>\n",
       "      <td>-0.190509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mito</th>\n",
       "      <td>0.577881</td>\n",
       "      <td>1.381192</td>\n",
       "      <td>1.271478</td>\n",
       "      <td>9.76</td>\n",
       "      <td>-0.803311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER</th>\n",
       "      <td>1.040071</td>\n",
       "      <td>0.828983</td>\n",
       "      <td>0.83908</td>\n",
       "      <td>9.32</td>\n",
       "      <td>0.211089</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     odds ratio restComp_oddsratio anyComp_oddsratio top-ratio       dif\n",
       "DNA    2.187816            0.64314          1.025543       9.1  1.544676\n",
       "RNA    0.952168           1.004431          0.993456      9.76 -0.052264\n",
       "AGP    0.851337           1.041846          0.992218     10.64 -0.190509\n",
       "Mito   0.577881           1.381192          1.271478      9.76 -0.803311\n",
       "ER     1.040071           0.828983           0.83908      9.32  0.211089"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f64e3779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds ratio</th>\n",
       "      <th>restComp_oddsratio</th>\n",
       "      <th>anyComp_oddsratio</th>\n",
       "      <th>top-ratio</th>\n",
       "      <th>dif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>2.167224</td>\n",
       "      <td>0.855631</td>\n",
       "      <td>1.196811</td>\n",
       "      <td>11.51</td>\n",
       "      <td>1.311593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA</th>\n",
       "      <td>1.088565</td>\n",
       "      <td>1.041325</td>\n",
       "      <td>1.041498</td>\n",
       "      <td>12.17</td>\n",
       "      <td>0.04724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGP</th>\n",
       "      <td>0.769428</td>\n",
       "      <td>1.127649</td>\n",
       "      <td>1.035857</td>\n",
       "      <td>13.05</td>\n",
       "      <td>-0.358221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mito</th>\n",
       "      <td>0.687255</td>\n",
       "      <td>1.219174</td>\n",
       "      <td>1.196811</td>\n",
       "      <td>11.51</td>\n",
       "      <td>-0.531919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER</th>\n",
       "      <td>1.08498</td>\n",
       "      <td>0.961039</td>\n",
       "      <td>0.9819</td>\n",
       "      <td>11.62</td>\n",
       "      <td>0.123941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     odds ratio restComp_oddsratio anyComp_oddsratio top-ratio       dif\n",
       "DNA    2.167224           0.855631          1.196811     11.51  1.311593\n",
       "RNA    1.088565           1.041325          1.041498     12.17   0.04724\n",
       "AGP    0.769428           1.127649          1.035857     13.05 -0.358221\n",
       "Mito   0.687255           1.219174          1.196811     11.51 -0.531919\n",
       "ER      1.08498           0.961039            0.9819     11.62  0.123941"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dc0a08c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds ratio</th>\n",
       "      <th>restComp_oddsratio</th>\n",
       "      <th>anyComp_oddsratio</th>\n",
       "      <th>top-ratio</th>\n",
       "      <th>dif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>0.873695</td>\n",
       "      <td>1.629147</td>\n",
       "      <td>1.724173</td>\n",
       "      <td>18.2</td>\n",
       "      <td>-0.755452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA</th>\n",
       "      <td>1.197029</td>\n",
       "      <td>1.262254</td>\n",
       "      <td>1.609984</td>\n",
       "      <td>19.96</td>\n",
       "      <td>-0.065225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGP</th>\n",
       "      <td>0.967497</td>\n",
       "      <td>1.16015</td>\n",
       "      <td>1.818722</td>\n",
       "      <td>16.01</td>\n",
       "      <td>-0.192653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mito</th>\n",
       "      <td>0.979993</td>\n",
       "      <td>1.309076</td>\n",
       "      <td>1.444275</td>\n",
       "      <td>18.42</td>\n",
       "      <td>-0.329083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER</th>\n",
       "      <td>0.831579</td>\n",
       "      <td>1.236551</td>\n",
       "      <td>1.541405</td>\n",
       "      <td>20.61</td>\n",
       "      <td>-0.404972</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     odds ratio restComp_oddsratio anyComp_oddsratio top-ratio       dif\n",
       "DNA    0.873695           1.629147          1.724173      18.2 -0.755452\n",
       "RNA    1.197029           1.262254          1.609984     19.96 -0.065225\n",
       "AGP    0.967497            1.16015          1.818722     16.01 -0.192653\n",
       "Mito   0.979993           1.309076          1.444275     18.42 -0.329083\n",
       "ER     0.831579           1.236551          1.541405     20.61 -0.404972"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "15cba6ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds ratio</th>\n",
       "      <th>restComp_oddsratio</th>\n",
       "      <th>anyComp_oddsratio</th>\n",
       "      <th>top-ratio</th>\n",
       "      <th>dif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>0.793062</td>\n",
       "      <td>1.353643</td>\n",
       "      <td>1.419434</td>\n",
       "      <td>20.5</td>\n",
       "      <td>-0.560581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA</th>\n",
       "      <td>1.291008</td>\n",
       "      <td>1.317432</td>\n",
       "      <td>1.339849</td>\n",
       "      <td>20.83</td>\n",
       "      <td>-0.026424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGP</th>\n",
       "      <td>0.834736</td>\n",
       "      <td>1.322061</td>\n",
       "      <td>1.637884</td>\n",
       "      <td>18.75</td>\n",
       "      <td>-0.487325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mito</th>\n",
       "      <td>1.174962</td>\n",
       "      <td>1.359223</td>\n",
       "      <td>1.566091</td>\n",
       "      <td>20.72</td>\n",
       "      <td>-0.184261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ER</th>\n",
       "      <td>1.050249</td>\n",
       "      <td>1.112284</td>\n",
       "      <td>1.320451</td>\n",
       "      <td>22.92</td>\n",
       "      <td>-0.062034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     odds ratio restComp_oddsratio anyComp_oddsratio top-ratio       dif\n",
       "DNA    0.793062           1.353643          1.419434      20.5 -0.560581\n",
       "RNA    1.291008           1.317432          1.339849     20.83 -0.026424\n",
       "AGP    0.834736           1.322061          1.637884     18.75 -0.487325\n",
       "Mito   1.174962           1.359223          1.566091     20.72 -0.184261\n",
       "ER     1.050249           1.112284          1.320451     22.92 -0.062034"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "be4a7bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{llllll}\n",
      "\\toprule\n",
      "{} & odds ratio & restComp\\_oddsratio & anyComp\\_oddsratio & top-ratio &       dif \\\\\n",
      "\\midrule\n",
      "DNA  &   0.836969 &           1.218462 &          1.126352 &     19.96 & -0.381493 \\\\\n",
      "RNA  &   1.127582 &           1.136146 &          1.217852 &     21.93 & -0.008563 \\\\\n",
      "AGP  &   1.053636 &           1.057746 &          1.270029 &     17.32 &  -0.00411 \\\\\n",
      "Mito &   1.005369 &           1.282287 &          1.289183 &     20.18 & -0.276917 \\\\\n",
      "ER   &   0.890449 &           1.049286 &          1.177658 &     22.26 & -0.158838 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(table3.to_latex())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a98dcc6",
   "metadata": {},
   "source": [
    "## GO terms search for overlap of highly predictable genes (top 58)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87d28528",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mitochondri :  9.21 %      pvalue: 0.2   oddsratio: 1.58\n",
      "golgi :  4.9 %      pvalue: 0.53   oddsratio: 0.69\n",
      "membrane :  7.17 %      pvalue: 0.56   oddsratio: 1.22\n",
      "cytoskeleton-actin :  9.52 %      pvalue: 0.21   oddsratio: 1.59\n",
      "endoplasmic :  6.78 %      pvalue: 1.0   oddsratio: 1.02\n",
      "RNA_nucleoli :  4.96 %      pvalue: 0.22   oddsratio: 0.64\n",
      "DNA :  10.77 %      pvalue: 0.05   oddsratio: 1.95\n",
      "any_comps:  7.18 %      pvalue: 0.13   oddsratio: 2.5\n",
      "no comps:  3.0 %      pvalue: 0.13   oddsratio: 0.4\n"
     ]
    }
   ],
   "source": [
    "# top58=pd.read_csv('./top_58_common.txt',header=None)[0].tolist()\n",
    "# gene_cats_bpcc=pd.read_csv('./go_bp_cc_D2021_each_gene_cat.txt',delimiter='\\t')\n",
    "import sys\n",
    "sys.path.insert(0, '../utils/') \n",
    "from readProfiles import rename_affyprobe_to_genename\n",
    "from saveAsNewSheetToExistingFile import saveAsNewSheetToExistingFile\n",
    "top58=pd.read_csv('./source/top_59_atleast_topIn3.txt',header=None)[0].tolist()\n",
    "gene_cats_bpcc=pd.read_csv('./source/GO_bp_cc_mf_direct_intersection_782.txt',delimiter='\\t')\n",
    "\n",
    "# comps=['mitochondri','Golgi','membrane','cytoskeleton','actin','endoplasmic','RNA','nucleol','cell division','mitosis','mitotic','cell cycle']\n",
    "\n",
    "# for c in comps:\n",
    "#     gene_cats_bpcc[c]=gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).apply(lambda x: x if c in x else '')+\\\n",
    "#     gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).apply(lambda x: x[:x.find(c)].split('~')[-1]+x[x.find(c):].split('GO')[0] if c in x else '')\n",
    "comps=['mitochondri','golgi','membrane','cytoskeleton','actin','endoplasmic','rna','nucleol',\\\n",
    "       'cell division','mitosis','mitotic','cell cycle','cytokine','hormone']\n",
    "\n",
    "for c in comps:\n",
    "    gene_cats_bpcc[c]=gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('GO')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('GO')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_MF_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('GO')[0] for si in locations_of_substring(x,c)]) if c in x else '')\n",
    "    \n",
    "gene_cats_bpcc['RNA_nucleoli']=gene_cats_bpcc['rna']+gene_cats_bpcc['nucleol']\n",
    "gene_cats_bpcc['DNA']=gene_cats_bpcc['cell division']+gene_cats_bpcc['mitosis']+gene_cats_bpcc['mitotic']+gene_cats_bpcc['cell cycle']\n",
    "\n",
    "gene_cats_bpcc['cytoskeleton-actin']=gene_cats_bpcc['cytoskeleton']+gene_cats_bpcc['actin']\n",
    "\n",
    "# gene_cats_bpcc['RNA_nucleoli']=gene_cats_bpcc[' rna']+gene_cats_bpcc['nucleol']\n",
    "# gene_cats_bpcc['DNA']=gene_cats_bpcc['cell division']+gene_cats_bpcc['mitosis']+gene_cats_bpcc['mitotic']+gene_cats_bpcc['cell cycle']\n",
    "\n",
    "# gene_cats_bpcc['cytoskeleton-actin']=gene_cats_bpcc['cytoskeleton']+gene_cats_bpcc['actin']+\\\n",
    "# gene_cats_bpcc['golgi']+gene_cats_bpcc['membrane']+gene_cats_bpcc['cytokine']\n",
    "# gene_cats_bpcc['ER']=gene_cats_bpcc['endoplasmic']\n",
    "# gene_cats_bpcc['mitochondria']=gene_cats_bpcc['mitochondri']#+gene_cats_bpcc['hormone']\n",
    "\n",
    "\n",
    "\n",
    "# gene_cats_bpcc['RNA_nucleoli']=gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).apply(lambda x: x if 'RNA' in x or 'nucleoli' in x else '')+\\\n",
    "# gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).apply(lambda x: x if 'RNA' in x or 'nucleoli' in x else '')\n",
    "\n",
    "# gene_cats_bpcc['DNA']=gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).apply(lambda x: x if 'cell division' in x or 'mitosis' in x or 'mitotic' in x or 'cell cycle' in x else '')+\\\n",
    "# gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).apply(lambda x: x if 'cell division' in x or 'mitosis' in x or 'mitotic' in x or 'cell cycle' in x else '')\n",
    "\n",
    "# gene_cats_bpcc['DNA']=gene_cats_bpcc[c]=gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).apply(lambda x: x if 'cell division' in x or 'mitosis' in x or 'cell cycle' in x else '')+\\\n",
    "# gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).apply(lambda x: x if 'cell division' in x or 'mitosis' in x or 'cell cycle' in x else '')\n",
    "\n",
    "\n",
    "gene_cats_bpcc.loc[gene_cats_bpcc['ID'].isin(top58),'top58']=True\n",
    "\n",
    "\n",
    "\n",
    "#####################################\n",
    "comps2=['mitochondri','golgi','membrane','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from scipy.stats import fisher_exact\n",
    "top_bool=(gene_cats_bpcc['top58']==True).values\n",
    "\n",
    "table1=pd.DataFrame(index=comps2+['any comp','no comp'],columns=['Prevalence','p-value','odds ratio'])\n",
    "for co in comps2:\n",
    "    enr_ratio=gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc[co]!='')].shape[0]/\\\n",
    "    gene_cats_bpcc[(gene_cats_bpcc[co]!='')].shape[0]\n",
    "    comp_bool=(gene_cats_bpcc[co]!='').values\n",
    "    oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, comp_bool))\n",
    "#     print(co, ':',gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc[co]!='')].shape[0],\\\n",
    "#          ', ',gene_cats_bpcc[(gene_cats_bpcc[co]!='')].shape[0])\n",
    "\n",
    "    table1.loc[co,['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "    print(co,\": \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "    \n",
    "gene_cats_bpcc['any_comps']=''\n",
    "for co in comps2:\n",
    "    gene_cats_bpcc['any_comps']=gene_cats_bpcc['any_comps']+gene_cats_bpcc[co]\n",
    "     \n",
    "    \n",
    "enr_ratio=gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc['any_comps']!='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['any_comps']!='')].shape[0]\n",
    "oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['any_comps']!='').values))\n",
    "print(\"any_comps: \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "table1.loc['any comp',['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "\n",
    "\n",
    "enr_ratio=gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc['any_comps']=='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['any_comps']=='')].shape[0]\n",
    "oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['any_comps']=='').values))\n",
    "print(\"no comps: \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "table1.loc['no comp',['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "\n",
    "# print('num top (52)/total genes (782): ', np.round((52/782)*100,2),'%')\n",
    "\n",
    "source_data_add='../Figs_Source_Data.xlsx'\n",
    "if 0:\n",
    "    saveAsNewSheetToExistingFile(source_data_add,table1,'ExtendedData6')\n",
    "\n",
    "# gene_cats_bpcc.to_csv('./GO/go_bp_cc_D2021_each_gene_cat_completed.csv',index=False)\n",
    "# gene_cats_bpcc.to_csv('./source/GO_bp_cc_mf_direct_intersection_782_completed.csv',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "83ea53e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mitochondri           1.58\n",
       "golgi                 0.69\n",
       "membrane              1.22\n",
       "cytoskeleton-actin    1.59\n",
       "endoplasmic           1.02\n",
       "RNA_nucleoli          0.64\n",
       "DNA                   1.95\n",
       "any comp              2.50\n",
       "no comp               0.40\n",
       "Name: odds ratio, dtype: float64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table1['odds ratio'].astype(float).round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4d4a04f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.58, 0.69, 1.22, 1.59, 1.02, 0.64, 1.95, 2.5 , 0.4 ])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table1['odds ratio'].astype(float).round(2).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "561b4533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(52, 23)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc[gene_cats_bpcc['ID'].isin(top58)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a73a384f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(782, 23)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene_cats_bpcc.shape\n",
    "mitochondri :  9.21 %      pvalue: 0.2   oddsratio: 1.58\n",
    "golgi :  4.9 %      pvalue: 0.53   oddsratio: 0.69\n",
    "membrane :  7.17 %      pvalue: 0.56   oddsratio: 1.22\n",
    "cytoskeleton-actin :  9.52 %      pvalue: 0.21   oddsratio: 1.59\n",
    "endoplasmic :  6.78 %      pvalue: 1.0   oddsratio: 1.02\n",
    "RNA_nucleoli :  5.68 %      pvalue: 0.73   oddsratio: 0.81\n",
    "DNA :  10.77 %      pvalue: 0.05   oddsratio: 1.95\n",
    "any_comps:  7.27 %      pvalue: 0.12   oddsratio: 2.31\n",
    "no comps:  3.28 %      pvalue: 0.12   oddsratio: 0.43\n",
    "num top (52)/total genes (782):  6.65 %"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dac7ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../utils/') \n",
    "from readProfiles import rename_affyprobe_to_genename\n",
    "########### 1 ###########\n",
    "filename='../results/SingleGenePred_cpCategoryMap/cat_scores_maps.xlsx'\n",
    "saved_scores=pd.read_excel(filename, sheet_name=None)\n",
    "# which_ds_model='LUAD-9-MLP-ht'\n",
    "which_ds_model='LUAD-9-MLP-keras-ht'\n",
    "# which_ds_model='LUAD-9-lasso-ht'\n",
    "dfcats=saved_scores[which_ds_model].rename(columns={'Unnamed: 0':'ID'})\n",
    "dfcats=dfcats[dfcats.columns[~dfcats.isna().any()].tolist()]\n",
    "dfcats2,_=rename_affyprobe_to_genename(dfcats.set_index('ID').T,dfcats.ID.tolist())\n",
    "dfcats=dfcats2.T.reset_index()\n",
    "\n",
    "# saved_scores.keys()\n",
    "\n",
    "########### 2 ###########\n",
    "Channelss=['DNA','RNA','AGP','Mito','ER']\n",
    "Channelss_cats=['DNA|Nuclei_AreaShape','RNA','AGP|Cytoplasm_AreaShape|Cells_AreaShape','Mito','ER']\n",
    "# Channelss_cats=['DNA|Nuclei_AreaShape','RNA','AGP|Cytoplasm_AreaShape','Mito','ER']\n",
    "\n",
    "for ci in range(len(Channelss)):\n",
    "    dfcats['max_'+Channelss[ci]]=dfcats.loc[:,dfcats.columns.str.contains(Channelss_cats[ci])].max(axis=1)\n",
    "\n",
    "    \n",
    "dfcats['top_channel']=dfcats[['max_'+Channelss[ci] for ci in range(len(Channelss))]].idxmax(axis=\"columns\")\n",
    "########### 3 ###########\n",
    "# gene_cats_bpcc=pd.read_csv('./go_BP_CC_MF_DIRECT_921.txt',delimiter='\\t')\n",
    "gene_cats_bpcc=pd.read_csv('./source/GO_bp_cc_mf_direct_LUAD_976.txt',delimiter='\\t')\n",
    "comps=['mitochondri','golgi','membrane','cytoskeleton','actin','endoplasmic',' rna','nucleol',\\\n",
    "       'cell division','mitosis','mitotic','cell cycle','cytokine','hormone']\n",
    "#    UP_KW_BIOLOGICAL_PROCESS bad\n",
    "# UP_KW_CELLULAR_COMPONENT\n",
    "# UP_KW_MOLECULAR_FUNCTION #null\n",
    "\n",
    "# GOTERM_BP_DIRECT\n",
    "# GOTERM_CC_DIRECT\n",
    "# GOTERM_MF_DIRECT\n",
    "for c in comps:\n",
    "    gene_cats_bpcc[c]=gene_cats_bpcc['GOTERM_CC_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_MF_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "    gene_cats_bpcc['GOTERM_BP_DIRECT'].astype(str).str.lower().apply(lambda x:\\\n",
    "    ''.join([x[:si].split('~')[-1]+x[si:].split('go')[0] for si in locations_of_substring(x,c)]) if c in x else '')#+\\\n",
    "#     gene_cats_bpcc['UP_KW_BIOLOGICAL_PROCESS'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "#     gene_cats_bpcc['UP_KW_CELLULAR_COMPONENT'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')+\\\n",
    "#     gene_cats_bpcc['UP_KW_MOLECULAR_FUNCTION'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split('~')[-1]+x[si:].split('kw')[0] for si in locations_of_substring(x,c)]) if c in x else '')#+\\\n",
    "#     gene_cats_bpcc['UP_SEQ_FEATURE'].astype(str).str.lower().apply(lambda x:\\\n",
    "#     ''.join([x[:si].split(':')[-1]+','+x[si:].split(',')[0] for si in locations_of_substring(x,c)]) if c in x else '')\n",
    "    \n",
    "gene_cats_bpcc['RNA_nucleoli']=gene_cats_bpcc[' rna']+gene_cats_bpcc['nucleol']\n",
    "gene_cats_bpcc['DNA']=gene_cats_bpcc['cell division']+gene_cats_bpcc['mitosis']+gene_cats_bpcc['mitotic']+gene_cats_bpcc['cell cycle']\n",
    "\n",
    "gene_cats_bpcc['cytoskeleton-actin']=gene_cats_bpcc['cytoskeleton']+gene_cats_bpcc['actin']+\\\n",
    "gene_cats_bpcc['golgi']+gene_cats_bpcc['membrane']+gene_cats_bpcc['cytokine']\n",
    "# gene_cats_bpcc['ER']=gene_cats_bpcc['endoplasmic']\n",
    "gene_cats_bpcc['mitochondria']=gene_cats_bpcc['mitochondri']+gene_cats_bpcc['hormone']\n",
    "\n",
    "gene_cats_bpcc=pd.merge(gene_cats_bpcc,dfcats,how='inner',on='ID')\n",
    "\n",
    "#########################\n",
    "Channelss_dict={'DNA':'DNA','RNA_nucleoli':'RNA','cytoskeleton-actin':'AGP','mitochondria':'Mito','endoplasmic':'ER'}\n",
    "Chan_rev_dict = dict(zip(Channelss_dict.values(),Channelss_dict.keys()))\n",
    "\n",
    "comps2=['mitochondri','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "gene_cats_bpcc['any_comps']=''\n",
    "for co in comps2:\n",
    "    gene_cats_bpcc['any_comps']=gene_cats_bpcc['any_comps']+gene_cats_bpcc[co]\n",
    "        \n",
    "from sklearn.metrics import confusion_matrix\n",
    "from scipy.stats import fisher_exact\n",
    "# top_bool=(gene_cats_bpcc['top58']==True).values\n",
    "table2=pd.DataFrame(index=Channelss_dict.keys(),columns=Channelss_dict.values())\n",
    "table3=pd.DataFrame(index=Channelss,columns=['odds ratio','restComp_oddsratio','anyComp_oddsratio','top-ratio'])\n",
    "# table3=pd.DataFrame(index=Channelss,columns=['Prevalence','anyComp-Prevalence','noComp-Prevalence','top-ratio'])\n",
    "\n",
    "table=[]\n",
    "for c in Channelss:\n",
    "    print(c)\n",
    "    c_r=list(set(Channelss)-set([c]))\n",
    "#     gene_cats_bpcc_highP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.7)]\n",
    "#     gene_cats_bpcc_lowP=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]<0.1)]\n",
    "    \n",
    "#     low_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[:100].index\n",
    "#     high_ind=gene_cats_bpcc[['max_'+c]].sort_values(by='max_'+c)[-100:].index    \n",
    "#     gene_cats_bpcc_highP=gene_cats_bpcc.loc[high_ind].reset_index(drop=True)\n",
    "#     gene_cats_bpcc_lowP=gene_cats_bpcc.loc[low_ind].reset_index(drop=True)\n",
    "    \n",
    "#     gene_cats_bpcc_highP[]\n",
    "    \n",
    "#     gene_cats_bpcc2=gene_cats_bpcc[(gene_cats_bpcc['max_'+c]>0.6) | (gene_cats_bpcc['max_'+c]<0.1)].reset_index(drop=True)\n",
    "    gene_cats_bpcc2=gene_cats_bpcc.copy()\n",
    "#     top_bool=(gene_cats_bpcc2['max_'+c]>0.6).values\n",
    "#     top_bool=(gene_cats_bpcc['max_'+c]<0.1).values\n",
    "#     top_bool=(gene_cats_bpcc['max_'+c]>0.6).values\n",
    "#     top_bool=((gene_cats_bpcc['top_channel']=='max_'+c) &\\\n",
    "#               (gene_cats_bpcc['max_'+c]>0.4) &\\\n",
    "#              (gene_cats_bpcc[['max_'+cr for cr in c_r]].max(axis=1)<0.4)).values\n",
    "    top_bool=((gene_cats_bpcc['max_'+c]-gene_cats_bpcc[['max_'+cr for cr in c_r]].max(axis=1)>0.05) &\\\n",
    "             (gene_cats_bpcc['max_'+c]>0.3)).values    \n",
    "    \n",
    "#     index_of_max=gene_cats_bpcc[(gene_cats_bpcc['top_channel']=='max_'+c)].sort_values(by='max_'+c,ascending=False)\\\n",
    "#     .index[0:100].values\n",
    "#     top_bool=np.zeros(gene_cats_bpcc.shape[0],).astype(bool)\n",
    "#     top_bool[index_of_max]=True    \n",
    "    \n",
    "#     hhgj\n",
    "    n_top=sum(top_bool)\n",
    "    print(n_top)\n",
    "    \n",
    "    table1=pd.DataFrame(index=comps2+['any comp','no comp'],columns=['Prevalence','p-value','odds ratio'])\n",
    "    for co in comps2:\n",
    "#         ratio_h=gene_cats_bpcc_highP[(gene_cats_bpcc_highP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_highP.shape[0]\n",
    "\n",
    "#         ratio_l=gene_cats_bpcc_lowP[(gene_cats_bpcc_lowP[co]!='')].shape[0]/\\\n",
    "#         gene_cats_bpcc_lowP.shape[0]\n",
    "# #         fdsfsd\n",
    "#         print(c,'-',co,': ',ratio_h/ratio_l,gene_cats_bpcc_highP.shape[0],gene_cats_bpcc_lowP.shape[0])\n",
    "        \n",
    "        enr_ratio=gene_cats_bpcc2[top_bool & (gene_cats_bpcc2[co]!='')].shape[0]/\\\n",
    "        gene_cats_bpcc2[(gene_cats_bpcc2[co]!='')].shape[0]\n",
    "        comp_bool=(gene_cats_bpcc2[co]!='').values\n",
    "        oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, comp_bool))\n",
    "    #     print(co, ':',gene_cats_bpcc[(gene_cats_bpcc['top58']==True) & (gene_cats_bpcc[co]!='')].shape[0],\\\n",
    "    #          ', ',gene_cats_bpcc[(gene_cats_bpcc[co]!='')].shape[0])\n",
    "        \n",
    "        table1.loc[co,['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "#         print(co,\": \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "        table2.loc[co,c]=oddsratio\n",
    "         \n",
    "    \n",
    "    enr_ratio=gene_cats_bpcc[top_bool & (gene_cats_bpcc['any_comps']!='')].shape[0]/\\\n",
    "    gene_cats_bpcc[(gene_cats_bpcc['any_comps']!='')].shape[0]\n",
    "    any_oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['any_comps']!='').values))\n",
    "#     print(\"any_comps: \",np.round(enr_ratio*100,2),'%     ','pvalue:', np.round(pvalue,2),'  oddsratio:',np.round(oddsratio,2))\n",
    "    table1.loc['any comp',['Prevalence','p-value','odds ratio']]=enr_ratio*100,pvalue,oddsratio\n",
    "\n",
    "    nocomp_enr_ratio=gene_cats_bpcc[(top_bool) & (gene_cats_bpcc['any_comps']=='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['any_comps']=='')].shape[0]\n",
    "\n",
    "    comps2=['mitochondria','cytoskeleton-actin','endoplasmic','RNA_nucleoli','DNA']\n",
    "    comps2.remove(Chan_rev_dict[c])\n",
    "    gene_cats_bpcc['rest_comps']=''\n",
    "    for co in comps2:\n",
    "        gene_cats_bpcc['rest_comps']=gene_cats_bpcc['rest_comps']+gene_cats_bpcc[co]    \n",
    "    \n",
    "    rest_enr_ratio=gene_cats_bpcc[(top_bool) & (gene_cats_bpcc['rest_comps']=='')].shape[0]/\\\n",
    "gene_cats_bpcc[(gene_cats_bpcc['rest_comps']=='')].shape[0]    \n",
    "    rest_oddsratio, pvalue = fisher_exact(confusion_matrix(top_bool, (gene_cats_bpcc['rest_comps']!='').values))\n",
    "    \n",
    "    table.append(table1)\n",
    "#     print('num top ('+ str(n_top)+')/total genes (912): ', np.round((n_top/912)*100,2),'%')\n",
    "    table3.loc[c,['odds ratio','restComp_oddsratio','anyComp_oddsratio','top-ratio']]=\\\n",
    "    table1.loc[Chan_rev_dict[c],'odds ratio'],rest_oddsratio,\\\n",
    "    any_oddsratio,np.round((n_top/912)*100,2)\n",
    "    \n",
    "#     table3.loc[c,['Prevalence','anyComp-Prevalence','noComp-Prevalence','top-ratio']]=table1.loc[Chan_rev_dict[c],'Prevalence'],\\\n",
    "#     enr_ratio*100,nocomp_enr_ratio*100,np.round((n_top/912)*100,2)    \n",
    "    \n",
    "#     print(Chan_rev_dict[c],': ',table1.loc[Chan_rev_dict[c],['Prevalence']].values)\n",
    "    \n",
    "# # table3['dif']=table3['Prevalence']-table3['anyComp-Prevalence']\n",
    "table3['dif']=table3['odds ratio']-table3['restComp_oddsratio']\n",
    "# print(table3['dif'].min(),table3['dif'].sum())\n",
    "table3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b92904e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "3cf59eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_cats_bpcc2=gene_cats_bpcc[(gene_cats_bpcc['top_channel']=='max_'+c) &(gene_cats_bpcc['max_'+c]>0.4)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "a327a507",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "max_AGP     0.507761\n",
       "max_Mito    0.501436\n",
       "max_DNA     0.493772\n",
       "max_RNA     0.511070\n",
       "dtype: float64"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc2[['max_'+cr for cr in c_r]].mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "48f90520",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "max_DNA     0.475004\n",
       "max_RNA     0.502518\n",
       "max_AGP     0.491805\n",
       "max_Mito    0.507971\n",
       "max_ER      0.555582\n",
       "dtype: float64"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cats_bpcc2[['max_'+c for c in Channelss]].median(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22372921",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
